// ottheme-colors/token.common.js
module.exports = {
  colorSecondaryLighter: "rgba(35, 123, 152, 0.14)",
  colorMustard: "rgb(252, 207, 68)",
  colorDark: "rgb(51, 51, 51)",
  colorIcecream: "rgb(114, 201, 202)",
  colorOysterPink: "rgb(239, 212, 212)",
  opacityColor: "rgba(0, 0, 0, 0.12)",
  colorBrandRed: "rgb(218, 55, 67)",
  opacityColorDark: "rgba(0, 0, 0, 0.2)",
  colorLightGray: "rgb(236, 236, 236)",
  colorSecondaryDarker: "rgb(53, 98, 115)",
  starColor: "rgb(254, 171, 108)",
  colorTransparent: "rgba(255, 255, 255, 0.6)",
  colorMenu: "rgb(30, 30, 30)",
  colorPremiumAccessBookModalBackground: "rgb(166, 122, 211)",
  colorMistyRose: "rgb(255, 230, 232)",
  colorJuneberry: "rgb(53, 98, 115)",
  opacityColorLight: "rgba(0, 0, 0, 0.06)",
  colorSecondary: "rgb(35, 123, 152)",
  colorBrandBlue: "rgb(35, 123, 152)",
  colorMacadamia: "rgb(219, 217, 213)",
  colorLight: "rgb(153, 153, 153)",
  colorDarker: "rgb(0, 0, 0)",
  colorAlmond: "rgb(196, 191, 185)",
  colorLightSilver: "rgb(172, 172, 172)",
  colorMalibu: "rgb(108, 182, 245)",
  colorCashew: "rgb(243, 242, 241)",
  opacityColorDarker: "rgba(0, 0, 0, 0.5)",
  bodyBg: "rgb(247, 247, 247)",
  colorDarkGray: "rgb(135, 135, 135)",
  opacityColorDarkHex: "rgb(204, 204, 204)",
  colorEarlyGirl: "rgb(218, 55, 67)",
  colorBlackCherry: "rgb(129, 21, 32)",
  colorCaspianPink: "rgb(229, 101, 110)",
  colorBlueberry: "rgb(96, 178, 208)",
  colorMidGray: "rgb(204, 204, 204)",
  colorAlert: "rgb(211, 3, 3)",
  colorChalkboard: "rgb(51, 51, 51)",
  colorOffersPrimary: "rgb(59, 149, 104)",
  colorSilver: "rgb(102, 102, 102)",
  darkenPercent: "10%",
  colorPrimaryDarker: "rgb(129, 21, 32)",
  colorPrimary: "rgb(218, 55, 67)",
  opacityColorLightHex: "rgb(240, 240, 240)",
  colorGranite: "rgb(68, 68, 68)",
  contentBg: "rgb(247, 247, 247)",
  colorGifts: "rgb(174, 214, 215)",
  colorSuccess: "rgb(117, 185, 111)",
  colorProvincialPink: "rgb(254, 248, 240)",
  colorLogoSecondary: "rgb(255, 255, 255)",
  colorHighlightBlue: "rgb(238, 244, 251)",
  opacityColorWhite: "rgba(255, 255, 255, 0.3)",
  colorPlum: "rgb(170, 98, 134)",
  colorMenuLink: "rgba(96, 178, 208, 0.14)",
  colorHuckleberry: "rgb(177, 217, 231)",
  colorPumpkin: "rgb(239, 143, 77)",
  colorMedium: "rgb(102, 102, 102)",
  colorWhite: "rgb(255, 255, 255)",
};
