// otkit-typography-desktop/token.common.js
module.exports = {
  xxlargeBoldFontWeight: "bold",
  largeBoldFontWeight: "bold",
  xlargeBoldFontSize: "32px",
  xlargeBoldLineHeight: "40px",
  fontFamily: "'BrandonText', -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol'",
  smallMediumFontWeight: "500",
  xsmallMediumFontSize: "14px",
  xsmallMediumLineHeight: "20px",
  mediumBoldFontWeight: "bold",
  xsmallRegularFontWeight: "normal",
  smallRegularFontSize: "16px",
  xxlargeBoldFontSize: "48px",
  largeBoldFontSize: "24px",
  smallRegularLineHeight: "24px",
  xxlargeBoldLineHeight: "56px",
  largeBoldLineHeight: "32px",
  fontWeightMedium: "500",
  smallMediumFontSize: "16px",
  smallMediumLineHeight: "24px",
  mediumBoldFontSize: "18px",
  mediumBoldLineHeight: "24px",
  xsmallRegularFontSize: "14px",
  largeMediumFontWeight: "500",
  xsmallRegularLineHeight: "20px",
  smallBoldFontWeight: "bold",
  xlargeBoldFontWeight: "bold",
  largeMediumFontSize: "24px",
  largeMediumLineHeight: "32px",
  fontWeightBold: "bold",
  fontFamilyBrand: "BrandonText",
  xsmallMediumFontWeight: "500",
  smallBoldFontSize: "16px",
  fontWeightNormal: "normal",
  smallBoldLineHeight: "24px",
  smallRegularFontWeight: "normal",
};
