// gca-react-components/src/components/PublishButton/PublishButton.js
/**
 * PublishButton wraps a gca-react-components Button with a default or transiently displayed success
 * message to the left.
 */
import React from 'react';
import PropTypes from 'prop-types';
import { Button } from '../Button/Button';
import { SuccessIcon } from '../../icons/SuccessIcon';
import { FlashMessage } from '../FlashMessage/FlashMessage';
import './PublishButton.scss';

export function PublishButton(props) {
  const button = (
    <Button
      className="gca-button"
      theme="publish"
      onClick={props.onClick}
      busy={props.busy || false}
      disabled={props.disabled || false}
      htmlType={props.htmlType}
    >
      {props.buttonText}
    </Button>
  );

  const successMessage = (
    <span className="success">
      <SuccessIcon className="success-icon" />
      {props.successText}
    </span>
  );

  const defaultMessage = <span>{props.restingText}</span>;

  const flashMessage = (
    <FlashMessage
      activeContent={successMessage}
      passiveContent={defaultMessage}
      activeDuration={props.successDuration}
      active={props.success}
    />
  );

  return (
    <span className="gca-publish-button">
      {flashMessage}
      {button}
    </span>
  );
}

PublishButton.propTypes = {
  restingText: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
  buttonText: PropTypes.string.isRequired,
  successText: PropTypes.string.isRequired,
  onClick: PropTypes.func,
  busy: PropTypes.bool,
  disabled: PropTypes.bool,
  success: PropTypes.bool,
  successDuration: PropTypes.number,
  htmlType: PropTypes.string,
};

PublishButton.defaultProps = {
  successDuration: 5000,
};
