// gc-microapp-marketing/src/PromotedInventory/Insights/helpers/constants.js
// Reservation Status
/*
export const CANCELLED = 'CANCELLED';
export const CANCELLEDWEB = 'CANCELLEDWEB';
export const CANCELLEDRESTAURANT = 'CANCELLEDRESTAURANT';
export const CANCELLEDDISPUTED = 'CANCELLEDDISPUTED';
export const NOSHOW = 'NOSHOW';
export const NOSHOW2 = 'NO-SHOW';
export const DISPUTED = 'DISPUTED';
export const NOSHOWEXCUSED = 'NOSHOWEXCUSED';
export const NOTCONFIRMED = 'NOTCONFIRMED';
export const DONE = 'DONE';
export const ASSUMEDDONE = 'ASSUMEDDONE';
export const ASSUMEDSEATED = 'ASSUMEDSEATED';
export const SEATED = 'SEATED';
export const SEATEDDISPUTED = 'SEATEDDISPUTED';
export const ARRIVED = 'ARRIVED';
export const BOOKED = 'BOOKED';

export const CLEARED = 'CLEARED';
export const NOANSWER = 'NOANSWER';
export const PAGED2 = 'PAGED2';
export const PAGED1 = 'PAGED1';
export const PAGED3 = 'PAGED3';
export const BUSTABLE = 'BUSTABLE';
export const PARTIALLYSEATED = 'PARTIALLYSEATED';
export const PARTIALLYARRIVED = 'PARTIALLYARRIVED';
export const ALLARRIVED = 'ALLARRIVED';
export const ENTREE = 'ENTREE';
export const DESSERT = 'DESSERT';
export const APPETIZER = 'APPETIZER';
export const LATE = 'LATE';
export const PAID = 'PAID';
export const CONFIRMED = 'CONFIRMED';
export const PENDING = 'PENDING';
export const CHECKDROPPED = 'CHECKDROPPED';
export const LEFTMESSAGE = 'LEFTMESSAGE';
export const WRONGNUMBER = 'WRONGNUMBER';
export const TABLEASSIGNED = 'TABLEASSIGNED';

export const BREAKFAST = 'BREAKFAST';
export const LUNCH = 'LUNCH';
export const DINNER = 'DINNER';
export const LATENIGHT = 'LATENIGHT';

export const PHONE = 'PHONE';
export const WALKIN = 'WALKIN';
export const YOURWEBSITE = 'YOUR_WEBSITE';
export const OTNETWORK = 'OT_NETWORK';
export const WAITLIST = 'WAIT_LIST';
export const PROMOTEDINVENTORY = 'PROMOTED_INVENTORY';
export const OTPARTNERS = 'OT_PARTNERS';
export const ONLINEPENDING = 'ONLINE_PENDING';
*/

export const CAMPAIGN_BONUS_POINTS = 'BONUS_POINTS_COVERS';
export const CAMPAIGN_PROMOTED_RESULTS_TARGETED = 'PROMOTED_RESULTS_WITH_TARGETING_COVERS';
export const CAMPAIGN_PROMOTED_RESULTS = 'PROMOTED_RESULTS_COVERS';
export const CAMPAIGN_OFFERS = 'OFFER_COVERS';
export const CAMPAIGN_SPECIALS = 'SPECIAL_COVERS';
export const CAMPAIGN_CUSTOM_PROMOTIONS = 'CUSTOM_PROMOTIONS_COVERS';
export const CAMPAIGN_LOW_EXPOSURE = 'LOW_EXPOSURE_COVERS';
export const CAMPAIGN_STANDARD_EXPOSURE = 'STANDARD_EXPOSURE_COVERS';
export const CAMPAIGN_HIGH_EXPOSURE = 'HIGH_EXPOSURE_COVERS';


/*
// Subscriptions
export const REPORT_COVERSBYMONTH_ERB = 'MONTHLY-COVERS-INSIGHTS-ERB-OTC';
export const REPORT_COVERSBYMONTH_GCOTC = 'MONTHLY-COVERS-INSIGHTS-GC-OTC';
*/

// Grouped Covers
export const POP_COVERS = 'POP_COVERS';
export const YOUR_NETWORK_COVERS = 'YOUR_NETWORK_COVERS';
export const BOOKED_COVERS = 'BOOKED_COVERS';
export const CANCELLED_COVERS = 'CANCELLED_COVERS';
export const CHANNEL_DIRECT_COVERS = 'CHANNEL_DIRECT_COVERS';
export const CHANNEL_DISCOVERY_COVERS = 'CHANNEL_DISCOVERY_COVERS';
export const OPENTABLE_PARTNER_COVERS = 'OPENTABLE_PARTNER_COVERS';
export const OPENTABLE_PLATFORMS_COVERS = 'OPENTABLE_PLATFORMS_COVERS';
export const OPENTABLE_NETWORK_COVERS = 'OPENTABLE_NETWORK_COVERS';
export const OPENTABLE_COVERS = 'OPENTABLE_COVERS';
export const ONLINE_COVERS = 'ONLINE_COVERS';
export const NO_SHOW_COVERS = 'NO_SHOW_COVERS';
export const FINISHED_COVERS = 'FINISHED_COVERS';
export const ASSUMED_FINISHED_COVERS = 'ASSUMED_FINISHED_COVERS';
export const SEATED_COVERS = 'SEATED_COVERS';
export const WALKIN_COVERS = 'WALKIN_COVERS';
export const PHONE_COVERS = 'PHONE_COVERS';
export const ONLINE_PENDING_COVERS = 'ONLINE_PENDING_COVERS';
export const NODATA = 'NODATA';
export const TOTAL_SEATED_COVERS = 'TOTAL_SEATED_COVERS';

/*
// Product Type
export const PRODUCT_TYPE_GUESTCENTER = 'GUEST_CENTER';
export const PRODUCT_TYPE_GUESTCENTERCONNECT = 'GUEST_CENTER_CONNECT';

// Modals
export const MODAL_INFO = 'MODAL_INFO';
export const MODAL_ERROR = 'MODAL_ERROR';
export const MODAL_PRINT_RESERVATION = 'MODAL_PRINT_RESERVATION';
export const MODAL_GUEST_INFORMATION = 'MODAL_GUEST_INFORMATION';
export const MODAL_SHIFT_OCCUPANCY_HELP = 'MODAL_SHIFT_OCCUPANCY_HELP';

// Print Modals
export const PRINT_RESERVATION = 'PRINT_RESERVATION';
export const PRINT_REFERRAL = 'PRINT_REFERRAL';
export const PRINT_GUEST_FREQUENCY = 'PRINT_GUEST_FREQUENCY';

// Turn Time keys
export const TURNTIMES_TYPE_AVG_TURN_TIME = 'avg_turn_time';
export const TURNTIMES_TYPE_PARTY_SIZE = 'party_size';
export const TURNTIMES_TYPE_TYPE = 'type';
export const TURNTIMES_TYPE_PHONE = 'phone';
export const TURNTIMES_TYPE_WALKIN = 'walkin';
export const TURNTIMES_TYPE_WEB = 'web';
export const TURNTIMES_TYPE_OTNETWORK = 'opentable';
export const TURNTIMES_TYPE_RESTREF = 'your_network';
export const TURNTIMES_TYPE_1_2 = 'type1_2';
export const TURNTIMES_TYPE_3_4 = 'type3_4';
export const TURNTIMES_TYPE_5_6 = 'type5_6';
export const TURNTIMES_TYPE_7_8 = 'type7_8';
export const TURNTIMES_TYPE_9_10 = 'type9_10';
export const TURNTIMES_TYPE_11 = 'type11';

export const TURNTIMES_PARTYSIZE_1_2 = '1-2';
export const TURNTIMES_PARTYSIZE_3_4 = '3-4';
export const TURNTIMES_PARTYSIZE_5_6 = '5-6';
export const TURNTIMES_PARTYSIZE_7_8 = '7-8';
export const TURNTIMES_PARTYSIZE_9_10 = '9-10';
export const TURNTIMES_PARTYSIZE_11 = '11+';

// Shift Occupancy
export const SHIFT_OCCUPANCY_RESERVED_COVERS = 'SHIFT_OCCUPANCY_RESERVED_COVERS';
export const SHIFT_OCCUPANCY_WALKINS = 'SHIFT_OCCUPANCY_WALKINS';
export const SHIFT_OCCUPANCY_COVER_OTNETWORK = 'SHIFT_OCCUPANCY_COVER_OTNETWORK';
export const SHIFT_OCCUPANCY_COVER_WEBSITE = 'SHIFT_OCCUPANCY_COVER_WEBSITE';
export const SHIFT_OCCUPANCY_COVER_PHONE = 'SHIFT_OCCUPANCY_COVER_PHONE';
export const SHIFT_OCCUPANCY_PROMOTIONS = 'SHIFT_OCCUPANCY_PROMOTIONS';

export const SHIFT_OCCUPANCY_TYPE_DAY = 'DAY';
export const SHIFT_OCCUPANCY_TYPE_WEEK = 'WEEK';
export const SHIFT_OCCUPANCY_TYPE_SHIFT = 'SHIFT';

export const SHIFT_OCCUPANCY_VIEW_DAY = 'SHIFT_OCCUPANCY_VIEW_DAY';
export const SHIFT_OCCUPANCY_VIEW_WEEK = 'SHIFT_OCCUPANCY_VIEW_WEEK';
export const SHIFT_OCCUPANCY_VIEW_SHIFT = 'SHIFT_OCCUPANCY_VIEW_SHIFT';

export const SHIFT_OCCUPANCY_SUMMARY_TITLE_DAY = 'SHIFT_OCCUPANCY_SUMMARY_TITLE_DAY';
export const SHIFT_OCCUPANCY_SUMMARY_TITLE_WEEK = 'SHIFT_OCCUPANCY_SUMMARY_TITLE_WEEK';
export const SHIFT_OCCUPANCY_SUMMARY_TITLE_SHIFT = 'SHIFT_OCCUPANCY_SUMMARY_TITLE_SHIFT';

export const SHIFT_OCCUPANCY_TYPE_WALKINS = 'WALKINS';
export const SHIFT_OCCUPANCY_TYPE_OTNETWORK = 'OTNETWORK';
export const SHIFT_OCCUPANCY_TYPE_WEBSITE = 'WEBSITE';
export const SHIFT_OCCUPANCY_TYPE_PHONE = 'PHONE';
*/

// Common
export const COMMON_SECOND = 'second';
export const COMMON_MINUTE = 'minute';
export const COMMON_HOUR = 'hour';

// Custom Tags
export const DIETARY_NEEDS = 'DIETARY_NEEDS';
export const SEATING_REQUESTS = 'SEATING_REQUESTS';
export const SPECIAL_GUESTS = 'SPECIAL_GUESTS';
export const SPECIAL_EVENTS_AND_PROMOS = 'SPECIAL_EVENTS_AND_PROMOS';

// Referral Report
export const REFERRAL_ID = 'REFERRER_ID';
export const REFERRAL_REFERRER = 'REFERRER';
export const REFERRAL_FIRST_NAME = 'FIRST_NAME';
export const REFERRAL_LAST_NAME = 'LAST_NAME';
export const REFERRAL_COMPANY = 'BUSINESS_NAME';
export const REFERRAL_PHONE = 'PHONE';
export const REFERRAL_EMAIL = 'EMAIL';
export const REFERRAL_LOCATION = 'LOCATION';
export const REFERRAL_TOTAL_RESERVATIONS = 'TOTAL_RESERVATIONS';
export const REFERRAL_TOTAL_COVERS = 'TOTAL_COVERS';
export const REFERRAL_FIRST_TIME_COVERS = 'FIRST_TIME_COVERS';
export const REFERRAL_TRAVELER_COVERS = 'TRAVELER_COVERS';
export const REFERRAL_CANCELLED = 'CANCELLED_COVERS';
export const REFERRAL_NOSHOWS = 'NOSHOW_COVERS';
export const REFERRAL_UNGROUPED = 'UNGROUPED';
export const REFERRAL_UNDEFINED = 'UNDEFINED';
export const REFERRAL_SUMMARY_TOTAL_COVERS = 'SUMMARY_TOTAL_COVERS';
export const REFERRAL_SUMMARY_TOTAL_REFERRERS = 'SUMMARY_TOTAL_REFERRERS';
export const REFERRAL_SUMMARY_TOP_REFERRERS = 'SUMMARY_TOP_REFERRERS';
export const REFERRAL_SUMMARY_TOP_COMPANIES = 'SUMMARY_TOP_COMPANIES';

// Guests
export const GUESTS_FIRST_TIME = 'GUESTS_FIRST_TIME';
export const GUESTS_REPEAT = 'GUESTS_REPEAT';
export const GUESTS_OUTOFTOWN_GUESTS = 'GUESTS_OUTOFTOWN_GUESTS';
export const GUESTS_LOCAL_GUESTS = 'GUESTS_LOCAL_GUESTS';
export const GUESTS_EMPTY = 'GUESTS_EMPTY';

// Tags
export const TAG_PROFILE_VIP = 'GC:PROFILE/VIP';
export const TAG_DIET_VEGETARIAN = 'GC:DIET/VEGETARIAN';
export const TAG_DIET_VEGAN = 'GC:DIET/VEGAN';
export const TAG_DIET_PESCATARIAN = 'GC:DIET/PESCATARIAN';
export const TAG_DIET_NO_RED_MEAT = 'GC:DIET/NO_RED_MEAT';
export const TAG_SEATING_SMOKING = 'GC:SEATING/SMOKING';
export const TAG_SEATING_SMOKING_AREA = 'GC:SEATING/SMOKING_AREA';
export const TAG_PROFILE_EMPLOYEE = 'GC:PROFILE/EMPLOYEE';
export const TAG_PROFILE_FRIEND_OF_OWNER = 'GC:PROFILE/FRIEND_OF_OWNER';
export const TAG_PROFILE_FRIEND_OF_EMPLOYEE = 'GC:PROFILE/FRIEND_OF_EMPLOYEE';
export const TAG_PROFILE_INVESTOR = 'GC:PROFILE/INVESTOR';
export const TAG_DIET_ALLERGY = 'GC:DIET/ALLERGY';
export const TAG_DIET_GLUTEN_FREE = 'GC:DIET/GLUTEN_FREE';
export const TAG_SERVICE_ALERT_CHEF = 'GC:SERVICE/ALERT_CHEF';
export const TAG_SERVICE_ALERT_MANAGER = 'GC:SERVICE/ALERT_MANAGER';
export const TAG_PROFILE_REGULAR = 'GC:PROFILE/REGULAR';
export const TAG_PROFILE_BLOGGER = 'GC:PROFILE/BLOGGER';
export const TAG_PROFILE_CRITIC = 'GC:PROFILE/CRITIC';
export const TAG_SERVICE_BIRTHDAY = 'GC:SERVICE/BIRTHDAY';
export const TAG_SERVICE_ANNIVERSARY = 'GC:SERVICE/ANNIVERSARY';
export const TAG_SERVICE_BUSINESS_MEAL = 'GC:SERVICE/BUSINESS_MEAL';
export const TAG_SERVICE_DATE = 'GC:SERVICE/DATE';
export const TAG_SERVICE_SPECIAL_OCCASION = 'GC:SERVICE/SPECIAL_OCCASION';
export const TAG_SERVICE_GRADUATION = 'GC:SERVICE/GRADUATION';
export const TAG_SEATING_VIEW = 'GC:SEATING/VIEW';
export const TAG_SEATING_QUIET_TABLE = 'GC:SEATING/QUIET_TABLE';
export const TAG_SEATING_HIGHCHAIR = 'GC:SEATING/HIGHCHAIR';
export const TAG_SEATING_WHEELCHAIR = 'GC:SEATING/WHEELCHAIR';
export const TAG_SEATING_WINDOW = 'GC:SEATING/WINDOW';
export const TAG_SEATING_BOOTH = 'GC:SEATING/BOOTH';
export const TAG_SEATING_PATIO = 'GC:SEATING/PATIO';
export const TAG_SERVICE_PRE_THEATRE = 'GC:SERVICE/PRE-THEATRE';

// Table Types
export const TABLE_TYPE_BAR = 'BAR';
export const TABLE_TYPE_COUNTER = 'COUNTER';
export const TABLE_TYPE_HIGHTOP = 'HIGHTOP';
export const TABLE_TYPE_OUTDOOR = 'OUTDOOR';
export const TABLE_TYPE_REGULAR = 'REGULAR';

// Occasions
export const OCCASION_ANNIVERSARY = 'ANNIVERSARY';
export const OCCASION_BIRTHDAY = 'BIRTHDAY';
export const OCCASION_BUSINESS_MEAL = 'BUSINESS_MEAL';
export const OCCASION_CELEBRATION = 'CELEBRATION';
export const OCCASION_DATE_NIGHT = 'DATE';
export const OCCASION_SPECIAL_OCCASSION = 'SPECIAL_OCCASION';

// Sources
// @TODO: These consts need to be verified against the API return values
export const SOURCE_OT_DIRECT = 'DIRECT';
export const SOURCE_OT_DISCOVERY = 'DISCOVERY';
export const SOURCE_OT_NETWORK = 'NETWORK';
export const SOURCE_OT_PARTNERS = 'PARTNERS';
export const SOURCE_PHONE = 'PHONE';
export const SOURCE_PROMOTED_INVENTORY = 'PROMOTED_INVENTORY';
export const SOURCE_WAITLIST = 'WAITLIST';
export const SOURCE_WALKIN = 'WALKIN';
export const SOURCE_YOUR_WEBSITE = 'WEBSITE';

// Online Channel
export const ONLINE_CHANNEL_OT = 'OT';
export const ONLINE_CHANNEL_OT_PARRTNERS = 'OT_PARTNERS';

// Discovery Channel
export const CHANNEL_DIRECT = 'DIRECT';
export const CHANNEL_DISCOVERY = 'DISCOVERY';
export const CHANNEL_RESTREF = 'RESTREF';

// OpenTable Intent / Channel Categories - GR-1519
export const OPENTABLE_DISCOVERY = 'OT_DISCOVERY';
export const OPENTABLE_DIRECT = 'OT_DIRECT';
export const YOUR_NETWORK = 'YOUR_NETWORK';

// Discovery Covers Campaign Categories - GR-1519
export const LOW_EXPOSURE = 'LOW_EXPOSURE';
export const STANDARD_EXPOSURE = 'STANDARD_EXPOSURE';

// Common
export const COMMON_WALKIN = 'walkin';
export const COMMON_YOUR_NETWORK = 'your_network';
export const COMMON_OPENTABLE_NETWORK = 'opentable_network';
export const COMMON_PHONE = 'phone';

export const COMMON_BONUS_POINTS = 'bonus_points';
export const COMMON_CUSTOM_PROMOTIONS = 'custom_promotions';
export const COMMON_OFFER = 'offer';
export const COMMON_SPECIALS = 'specials';
export const COMMON_PROMOTED_RESULTS = 'promoted_results';
export const COMMON_PROMOTED_RESULTS_WITH_TARGETING = 'promoted_results_with_targeting';
export const COMMON_LOW_EXPOSURE = 'low_exposure';
export const COMMON_STANDARD_EXPOSURE = 'standard_exposure';

// Products
export const PRODUCT_RESERVATIONS = 'RESERVATIONS';
export const PRODUCT_GUEST_EXPORT = 'GUESTEXPORT';
export const PRODUCT_REFERRAL = 'REFERRAL';
export const PRODUCT_GUEST_FREQUENCY = 'GUESTFREQUENCY';

export const ACCOUNT_CONNECT = 'CONNECT';
export const ACCOUNT_GUESTCENTER = 'GUESTCENTER';
export const ACCOUNT_DEFAULT = 'DEFAULT';

// Columns width for tables
const paddingWidthOffset = 16;

export const TABLE_COLUMNS = {
  sizeS: 80 + paddingWidthOffset,
  sizeM: 128 + paddingWidthOffset,
  sizeL: 176 + paddingWidthOffset,
  sizeXL: 272 + paddingWidthOffset,
};
