// user-accounts/src/utils/trackingUtils.ts
import eventTracker from '@buffet/event-tracker';


let trackingCategory = 'User Accounts';

// rename trackingCategory and set "App Name" custom dimension
if (typeof window !== 'undefined' && window.gcScope) {
  const { gcScope } = window;

  if (gcScope.type === 'group') {
    trackingCategory = `Group ${trackingCategory}`;
  } else {
    trackingCategory = `Restaurant ${trackingCategory}`;
  }

  eventTracker.queue('set', 'dimension4', trackingCategory);
}

eventTracker.enable();

// Actions and Labels
const trackingEvents = {
  addUser: {
    action: 'add user',
    labels: {
      addButton: 'add new user',
      cancelButton: 'cancel',
      createButton: 'create',
      createTryButton: 'try create',
    },
  },
  customizeAccessTypes: {
    action: 'customize access type',
    labels: {
      cancelButton: 'cancel',
      customizeButton: 'customize',
      saveButton: 'save',
    },
  },
  editUser: {
    action: 'edit user',
    labels: {
      cancelButton: 'cancel',
      editButton: 'edit',
      saveButton: 'save',
      saveTryButton: 'try save',
    },
  },
  removeUser: {
    action: 'remove user',
    labels: {
      cancelButton: 'cancel',
      removeButton: 'remove',
      removeDialog: 'prompt',
    },
  },
  searchUsers: {
    action: 'search users',
    labels: {
      gotoPage: 'go to page',
      searchQuery: 'search query',
      sortField: 'sort',
    },
  },
};

export { eventTracker, trackingCategory, trackingEvents };
