// gc-microapp-login/src/common/api.ts
import {
  ChangePasswordRequest,
  LoginRequest,
  LoginResponse,
  RegisterPhoneRequest,
} from './types';

import { toJson } from './utils';

const postHeader = { 'Content-Type': 'application/json', 'X-Login-Version': 'v2' };

// ======= exported APIs ========================================

export function login(loginRequest: Partial<LoginRequest>): Promise<LoginResponse> {
  return fetch(
    '/gateway/api/account/login',
    {
      body: JSON.stringify(loginRequest),
      headers: { ...postHeader },
      method: 'POST',
    })
    .then(toJson)
}

export function resend2faCode(loginRequest: Partial<LoginRequest>): Promise<LoginResponse> {
  return fetch(
    '/gateway/api/account/resend-2fa-code',
    {
      body: JSON.stringify(loginRequest),
      headers: { ...postHeader },
      method: 'POST',
    })
    .then(toJson)
}

export function changePassword(changePasswordRequest: Partial<ChangePasswordRequest>): Promise<LoginResponse> {
  return fetch('/gateway/api/account/password/change', {
    body: JSON.stringify(changePasswordRequest),
    headers: { ...postHeader },
    method: 'POST',
  }).then(data => data.status === 200 ? Promise.resolve({}) : data.json());
}

export function resetPassword(email: Partial<string>): Promise<any> {
  const url = `/gateway/api/account/password/init/${email}`;
  return fetch(
    url,
    {
      headers: { ...postHeader },
      method: 'POST',
    })
}

export function registerPhone(registerPhoneRequest: Partial<RegisterPhoneRequest>): Promise<LoginResponse> {
  const url = `/gateway/api/account/register-two-factor`;
  return fetch(
    url,
    {
      body: JSON.stringify(registerPhoneRequest),
      headers: { ...postHeader },
      method: 'POST',
    })
    .then(toJson)
}
