// gc-microapp-bistro/src/common/DataDisplay/DataTableDisplay.tsx
import * as React from 'react';
import ReactTable from 'react-table';
import 'react-table/react-table.css';

import strings from '../../strings';

import './DataDisplay.scss';
import './DefaultTableStyle.override.scss';
import './OtTableStyle.override.scss';

export interface Props {
  data: any;
}

export default class DataTableDisplay extends React.Component<Props, any> {
  constructor(props: Props) {
    super(props);

    this.state = {
      status: null,
    };
  }

  public render() {
    const { data } = this.props;
    const columns = [
      {
        Header: 'RID',
        accessor: 'rid',
        id: 'rid',
        width: 80,
      },
      {
        Header: 'Restaurant',
        accessor: 'entityInfo.restaurantName',
        id: 'restaurantName',
        width: 256,
      },
      {
        Header: 'Contract Owner',
        accessor: (row: any) => `${row.contractOwner.firstName} ${row.contractOwner.lastName}`,
        id: 'contractOwner',
        width: 128,
      },
      {
        Header: 'Client Contact',
        accessor: (row: any) => `${row.customerInfo.firstName} ${row.customerInfo.lastName}`,
        id: 'customer',
        width: 128,
      },
      {
        Header: 'Agreement Type',
        accessor: (row: any) => `${this.getAgreementType(row.type)}`,
        id: 'agreementType',
        width: 160,
      },
      {
        Header: 'State',
        accessor: (row: any) => `${this.getAgreementState(row.agreementState)}`,
        id: 'agreementState',
        width: 128,
      },
      {
        Cell: (e: any) => `${this.getDate(e.value)}`,
        Header: 'Created Date',
        accessor: 'createdDate',
        id: 'createdDate',
        width: 128,
      },
      {
        Cell: (e: any) => `${this.getDate(e.value)}`,
        Header: 'Signed Date',
        accessor: 'signedDate',
        id: 'signedDate',
        width: 128,
      },
      {
        // tslint:disable-next-line:max-line-length
        Cell: (e: any) => (
          <a target="_blank" href={`/admin/user-agreements/${e.value}/view`}>
            View{' '}
          </a>
        ),
        Header: 'View Agreement',
        accessor: 'agreementId',
        id: 'viewAgreemnt',
        sortable: false,
        width: 128,
      },
    ];

    return (
      <div className="data-container">
        <h3>User Agreements</h3>
        <ReactTable
          data={data}
          columns={columns}
          sortable={true}
          showPagination={false}
          resizable={true}
          defaultSorted={[
            {
              desc: true,
              id: 'createdDate',
            },
          ]}
        />
      </div>
    );
  }

  private getAgreementState(state: string): string {
    return strings[`agreement.state.${state}`];
  }

  private getAgreementType(type: string): string {
    return strings[`agreement.type.${type}`];
  }

  private getDate(dateStr: string): string {
    let formatDate = '';
    if (dateStr && dateStr.length > 0) {
      const dateObj: Date = new Date(dateStr);
      formatDate = `${dateObj.getMonth() + 1}-${dateObj.getDate()}-${dateObj.getFullYear()}`;
    }
    return formatDate;
  }
}
