// gc-microapp-reviews/src/Reviews/AllReviews/ReviewBreakdown/ReviewBreakdown.tsx
import React from 'react';
import { FormattedMessage } from 'react-intl';

import { Rating } from '../../../store/allReviews/allReviews';

import './ReviewBreakdown.scss';

interface Props {
  ratings: Rating;
}

interface ReviewBreakdownItemProps {
  ratingId: string;
  ratingValue?: number;
}

class ReviewBreakdownItem extends React.Component<ReviewBreakdownItemProps> {
  public static defaultProps = {
    ratingValue: '0',
  };

  public render() {
    const { ratingId, ratingValue } = this.props;
    return (
      <li>
        <FormattedMessage tagName="h5" id={`reviews.summary.ratingLabels.${ratingId}`} /> <p>{ratingValue || '-'}</p>
      </li>
    );
  }
}

function ReviewBreakdown({ ratings: { Food, Service, Ambience, Value, Noise } }: Props) {
  return (
    <ul className="gca-review-breakdown">
      <ReviewBreakdownItem ratingId="food" ratingValue={Food} />
      <ReviewBreakdownItem ratingId="service" ratingValue={Service} />
      <ReviewBreakdownItem ratingId="ambience" ratingValue={Ambience} />
      <ReviewBreakdownItem ratingId="value" ratingValue={Value} />
      <ReviewBreakdownItem ratingId="noise" ratingValue={Noise} />
    </ul>
  );
}

export default ReviewBreakdown;
