// harness-smartling-client/lib/client/SmartlingClient.js
"use strict";function _classCallCheck(instance, Constructor) {if (!(instance instanceof Constructor)) {throw new TypeError("Cannot call a class as a function");}}module.exports = function (SmartlingAuthenticateService,
SmartlingDownloadOptions, SmartlingDownloadService, SmartlingProcessDownloadResponse,
SmartlingUploadOptions, SmartlingUploadService,
SmartlingImportService, SmartlingImportOptions) {var
  SmartlingClient = function () {
    function SmartlingClient(userIdentifier, userSecret, projectId) {_classCallCheck(this, SmartlingClient);
      this.userIdentifier = userIdentifier;
      this.userSecret = userSecret;
      this.projectId = projectId;
      this.accessToken = null;
    }SmartlingClient.prototype.

    authenticate = function authenticate() {var _this = this;
      return SmartlingAuthenticateService.fetch(this.userIdentifier, this.userSecret).
      then(function (response) {
        _this.accessToken = response.accessToken;
        return response;
      });
    };SmartlingClient.prototype.

    download = function download(options) {var _this2 = this;
      var criteria = new SmartlingDownloadOptions({
        projectId: this.projectId,
        fileUri: options.fileUri,
        localeIds: options.localeIds,
        fileNameMode: options.fileNameMode,
        localeMode: options.localeMode,
        retrievalType: options.retrievalType });


      return this.authenticate().
      then(this._setAccessToken.bind(this)).
      then(function () {
        criteria.accessToken = _this2.accessToken;

        return SmartlingDownloadService.fetch(criteria).
        then(function (downloadBuffer) {// eslint-disable-line
          return SmartlingProcessDownloadResponse.process(downloadBuffer, criteria, options.saveTo);
        });
      });
    };SmartlingClient.prototype.

    upload = function upload(options) {var _this3 = this;
      var criteria = new SmartlingUploadOptions({
        projectId: this.projectId,
        fileUri: options.fileUri,
        fileType: options.fileType,
        content: options.content,
        contentType: options.contentType });


      return this.authenticate().
      then(this._setAccessToken.bind(this)).
      then(function () {
        criteria.accessToken = _this3.accessToken;

        return SmartlingUploadService.send(criteria);
      });
    };SmartlingClient.prototype.

    import = function _import(options) {var _this4 = this;
      var criteria = new SmartlingImportOptions({
        projectId: this.projectId,
        localeId: options.localeId,
        fileUri: options.fileUri,
        fileType: options.fileType,
        content: options.content,
        contentType: options.contentType,
        overwrite: options.overwrite,
        translationState: options.translationState });


      return this.authenticate().
      then(this._setAccessToken.bind(this)).
      then(function () {
        criteria.accessToken = _this4.accessToken;

        return SmartlingImportService.send(criteria);
      });
    };SmartlingClient.prototype.

    _setAccessToken = function _setAccessToken(authenticateResponse) {
      this.accessToken = authenticateResponse.accessToken;
    };return SmartlingClient;}();


  return SmartlingClient;
};
