// gc-microapp-marketing/src/store/insights/mockCoverTypesDone.js
export const mockCoverTypesDoneAction = {
  type: 'INSIGHTS/FETCH_COVER_TYPES_DONE',
  payload: {
    params: {
      endDate: '2019-05-02',
      startDate: '2019-04-26',
    },
    result: {
      domain: ['PAID250-438647', 'PAID250-438644'],
      range: {
        'PAID250-438647': {
          domain: ['2019-02-03', '2019-02-02'],
          range: {
            '2019-02-03': {
              campaign_id: 438647,
              cover_type: 'PAID250',
              date_actual: '2019-02-03',
              covers: 2,
            },
            '2019-02-02': {
              campaign_id: 438647,
              cover_type: 'PAID250',
              date_actual: '2019-02-02',
              covers: 2,
            },
          },
        },
        'PAID250-438644': {
          domain: ['2019-02-03'],
          range: {
            '2019-02-03': {
              campaign_id: 438644,
              cover_type: 'PAID250',
              date_actual: '2019-02-03',
              covers: 2,
            },
          },
        },
      },
    },
  },
};
