// @buffet/user-agreement/src/FooterUserAgreement.tsx
/*istanbul ignore file */
import React from 'react';
import { TransitionableFooter } from './Footer/TransitionableFooter';
import { UserAgreement, UserAgreementProps } from './UserAgreement';
import styles from './FooterUserAgreement.css';
import { UserAgreementInteractionProps } from './UserAgreementUnsigned/UserAgreementUnsigned';

export interface ViewProps extends UserAgreementProps, UserAgreementInteractionProps {
  show: boolean; // to render the UserAgreement component
  persistentFooter?: boolean; // to render the empty footer before the UserAgreement
  className?: string;
}

export class FooterUserAgreement extends React.Component<ViewProps> {
  render() {
    const {
      show,
      persistentFooter = false,
      agreementType,
      context,
      contextName,
      onAgreeSuccess,
      pricing,
      onInitialized,
      children,
      className,
      onAgreeIntent,
      onAgreeConfirm,
      onViewAgreement,
      onHideAgreement
    } = this.props;

    const hasChildren = !!children;

    return (
      <TransitionableFooter
        show={show || persistentFooter}
        className={`${styles.footer} ${className} ${hasChildren ? styles.hasChildren : styles.noChildren}`}
      >
        {hasChildren && <div className={styles.customContent}>{children}</div>}
        {(show || !persistentFooter) && (
          <div className={persistentFooter && styles.animateAgreementIn}>
            <UserAgreement
              agreementType={agreementType}
              context={context}
              contextName={contextName}
              pricing={pricing}
              onInitialized={onInitialized}
              onAgreeSuccess={onAgreeSuccess}
              onAgreeIntent={onAgreeIntent}
              onAgreeConfirm={onAgreeConfirm}
              onViewAgreement={onViewAgreement}
              onHideAgreement={onHideAgreement}
            />
          </div>
        )}
      </TransitionableFooter>
    );
  }
}
