// gc-microapp-bistro/src/common/DataDisplay/DataPartnerDisplay.tsx
import * as React from 'react';
import ReactTable from 'react-table';
import 'react-table/react-table.css';

import strings from '../../strings';

import './DataDisplay.scss';
import './DefaultTableStyle.override.scss';
import './OtTableStyle.override.scss';

export interface Props {
  data: any;
}

export default class DataPartnerDisplay extends React.Component<Props, any> {
  constructor(props: Props) {
    super(props);

    this.state = {
      status: null,
    };
  }

  public render() {
    const { data } = this.props;
    const columns = [
      {
        Header: 'Partner RID',
        accessor: 'item.id',
        id: 'partnerId',
        width: 128,
      },
      {
        Cell: (e: any) => `${this.getPartnerName(e.value)}`,
        Header: 'Partner Name',
        accessor: 'partnerId',
        id: 'partnerName',
        width: 256,
      },
      {
        Header: 'Restaurant Name',
        accessor: 'item.name',
        id: 'name',
        width: 256,
      },
      {
        Header: 'Country',
        accessor: 'item.address.country',
        id: 'country',
        width: 128,
      },
      {
        Header: 'Status Timestamp',
        accessor: 'statusTimestamp',
        id: 'statusTimestamp',
        width: 256,
      },
      {
        Header: 'Status',
        accessor: 'status',
        id: 'status',
        width: 128,
      },
      {
        Header: 'Status Details',
        accessor: 'statusDetails',
        id: 'statusDetails',
        width: 512,
      },
    ];

    return (
      <div className="data-container">
        <h3>Partner List</h3>
        <ReactTable
          data={data.items}
          columns={columns}
          sortable={true}
          showPagination={true}
          resizable={true}
        />
      </div>
    );
  }

  private getPartnerName(partnerId: string): string {
    return strings[`partnerships.partnerName.${partnerId}`];
  }
}
