// @buffet/modal/src/Content/Footer.tsx
import { Button } from '@buffet/button';
import classnames from 'classnames/bind';
import React from 'react';

import { ModalType } from '../types';

import styles from './Footer.css';


const cx = classnames.bind(styles);

export interface ModalFooterProps {
  acknowledgeLabel?: string;
  onAcknowledge?: () => void;
  dismissLabel?: string;
  onDismiss?: () => void;
  modalType?: ModalType;
}

/* istanbul ignore next */
const noop = () => {};

const ModalFooter = ({
  acknowledgeLabel,
  onAcknowledge = noop,
  dismissLabel,
  onDismiss = noop,
  modalType,
}: ModalFooterProps) => (
  <footer className={cx('footer', modalType === ModalType.fullscreen && 'fullscreen')}>
    { dismissLabel && (
      <Button type="button" theme="bare" size="small" htmltype="button" onClick={onDismiss}>{dismissLabel}</Button>
    )}
    { acknowledgeLabel && (
      <Button type="button" theme="primary" size="small" htmltype="button" onClick={onAcknowledge}>{acknowledgeLabel}</Button>
    )}
  </footer>
);

export default ModalFooter;
