// gc-microapp-marketing/src/store/privatedining/privateDiningForm.ts
import { createSelector } from 'reselect';
import { photo } from './mappers';
import { privateDiningSelector } from './privatedining';
import { PrivateDining, PrivateDiningData, Room } from './privatedining.types';

const roomDefault = (room = {}, index: number): Room => ({
  active: false,
  imageName: '',
  imageURL: '',
  listingOrder: index + 1,
  photo,
  privateRoomImageId: 0,
  roomChangeMask: 0,
  roomDescription: '',
  roomId: 0,
  roomName: '',
  seatedSize: 0,
  standingSize: 0,
  ...room,
});

export const privateDiningFormDefaults = (partialPrivateDiningForm: Partial<PrivateDiningData>): PrivateDiningData => {
  const rooms = partialPrivateDiningForm.rooms || [];

  return {
    draftSubmitted: true,
    enablePrivateDining: false,
    hasPrivateParty: false,
    isPDDraft: 0,
    languageID: 0,
    largestCapacity: 0,
    locale: 'en-US',
    modified: '',
    newPremiumRequest: true,
    partyLongDescription: '',
    partyShortDescription: '',
    pdChangeMask: 0,
    photo,
    premiumEnabled: true,
    privateDiningMenu: '',
    privateDiningThumbnail: '',
    privateDiningThumbnailURL: '',
    privatePartyContact: '',
    privatePartyEmail: '',
    privatePartyPhone: '',
    privatePartyPhoneExt: '',
    publishPrivateDining: false,
    requestId: null,
    rid: 0,
    submittedBy: '',
    thumbnailImageID: 0,
    ...partialPrivateDiningForm,
    rooms: [0, 1, 2, 3, 4].map(i => roomDefault(rooms[i], i)),
  };
};

export const initialValuesSelector = createSelector(privateDiningSelector, privateDining => {
  if (privateDining === null) {
    return privateDiningFormDefaults({});
  }

  return privateDiningFormDefaults(privateDining);
});
