// gc-microapp-marketing/src/PromotedInventory/CampaignsList/campaign-actions.ts
import clipboard from 'clipboard-polyfill';

import { CampaignType, ModalType, OpenModalPayload, PatchActionPayload } from '../../store/campaigns/campaigns.types';

const campaignActions = {
  handleCopyLink: (campaignId: string) => {
    const { origin, pathname } = window.location;
    clipboard.writeText(`${origin}${pathname}#${campaignId}`);
  },
  handleEditBudget: (campaign: CampaignType, openEditBudgetModal: (payload: CampaignType) => void) => {
    openEditBudgetModal(campaign);
  },
  handleEnd: (campaign: CampaignType, openModal: (payload: OpenModalPayload) => void) => {
    openModal({
      campaign,
      type: ModalType.End,
    });
  },
  handlePause: (campaign: CampaignType, openModal: (payload: OpenModalPayload) => void) => {
    openModal({
      campaign,
      type: ModalType.Pause,
    });
  },
  handleResume: (campaignId: string, patchCampaignRequest: (payload: PatchActionPayload) => void) => {
    patchCampaignRequest({
      campaignId,
      refetchCampaigns: true,
      state: 'live',
    });
  },
};

export default campaignActions;
