// @buffet/radio-button/src/RadioButtonGroup.tsx
import React from 'react';
import { SimpleRadioButtonGroup } from './SimpleRadioButtonGroup';
import { RadioButton } from './RadioButton';

interface Datum {
  value: string;
  label: React.ReactNode;
  disabled?: boolean;
}
interface Props {
  childProvider: Datum[];
  onChange(value: string): void;
  selectedValue: string;
  className?: string;
  radioOptionClassName?: string;
}

export class RadioButtonGroup extends React.Component<Props> {
  render() {
    const { childProvider, onChange, selectedValue, className, radioOptionClassName } = this.props;
    return (
      <SimpleRadioButtonGroup className={className} onChange={onChange} selectedValue={selectedValue}>
        {childProvider.map(childProps => {
          const mergedProps = {
            ...childProps,
            onChange,
            className: radioOptionClassName,
            checked: childProps.value === selectedValue,
            key: `rbg_${childProps.value}`
          };
          return <RadioButton {...mergedProps} />;
        })}
      </SimpleRadioButtonGroup>
    );
  }
}
