// damper/src/components/Bar/Bar.story.js
import React from 'react';
import { boolean, number } from '@storybook/addon-knobs';
import { Bar } from '../Bar';
import theme from '../../theme/defaultTheme';

class BarStory extends React.Component {
  render() {
    const valueOption = {
      range: true,
      min: 0,
      max: 800,
      step: 1,
    };

    const roundingOption = {
      range: true,
      min: 0,
      max: 100,
      step: 1,
    };

    const widthOption = {
      range: true,
      min: 0,
      max: 100,
      step: 1,
    };

    const isMounted = boolean('isMounted', true);
    const isAnimated = boolean('isAnimated', true);

    const value = number('Value', 200, valueOption);
    const rounding = number('Rounding', 0, roundingOption);
    const width = number('Width', 60, widthOption);

    let customisedTheme = theme;

    if (rounding > 0) {
      customisedTheme = {
        ...customisedTheme,
        bars: {
          ...customisedTheme.bars,
          default: {
            ...customisedTheme.bars.default,
            rounding,
          },
        },
      };
    }

    customisedTheme = {
      ...customisedTheme,
      bars: {
        ...customisedTheme.bars,
        default: {
          ...customisedTheme.bars.default,
          width,
        },
      },
    };

    const graphHeight = 450;

    return (
      <article className="markdown-body">
        <h1>Bar Component</h1>
        <svg width="100%" height={graphHeight}>
          {isMounted && (
            <Bar
              actions={{
                onClick: () => {},
                onMouseOver: () => {},
                onMouseOut: () => {},
              }}
              id="storybook-test"
              isAnimated={isAnimated}
              scale={(d) => d}
              theme={customisedTheme}
              value={value}
              xOffset={width / 2}
              yOffset={graphHeight}
            />
          )}
        </svg>
        <hr />
        <p style={{ margin: '24px 0' }}>
          The Bar component is heavily used in bar graphs. It requires y-alignment within the component
          that&#39;s using it.
        </p>
        <p>
        Current settings for the previewed bar are;<br />
          Width: {width}px<br />
          Height: {value}px (from value)<br />
          Rounding: {rounding}px<br />
          isAnimated: {`${isAnimated}`}<br />
        </p>
      </article>
    );
  }
}

export { BarStory };
