// UICard/UICard.tsx
import React from 'react';

import './UICard.scss';

export interface UICardProps {
  imageComponent: any;
  cta: string; // cta text for url link
  ctaUrl: string;
  ctaOpenTab?: boolean; // open tab when click cta link
  description: string;
  linkable?: boolean; // is the link clickable
  title: string; // title
  clickCallback?: () => void;
}

export class UICard extends React.Component<UICardProps> {
  public render() {
    const { imageComponent, cta, description, linkable = true, title } = this.props;

    const enabledClass = linkable ? 'enabled' : '';

    return (
      <div className={`uicard ${enabledClass}`} data-qa="uicard" {...(linkable ? { onClick: this.handleClick } : {})}>
        <div className="uicard-content">
          <div className="card-illustration">{imageComponent}</div>
          <h3 className="card-title">{title}</h3>
          <div className="card-description">{description}</div>
          <div className="card-call-to-action">{cta}</div>
        </div>
      </div>
    );
  }

  private handleClick = () => {
    window.open(this.props.ctaUrl, this.props.ctaOpenTab ? '_blank' : '_self');
    if (this.props.clickCallback) { this.props.clickCallback(); }
  };
}
