// user-agreements/src/ContractOwner/ContractOwner.tsx
import React from 'react';
import { FormattedMessage, InjectedIntlProps, injectIntl } from 'react-intl';
import { connect } from 'react-redux';

import { State } from '../store/types';
import { actions, selectors } from '../store/userAgreements/userAgreements';
import { ContactInfo } from '../store/userAgreements/userAgreements.types';
import './ContractOwner.scss';

interface SP {
  contractOwner: ContactInfo;
}

interface DP {
  emailContractOwner: () => void;
}

type Props = SP & DP & InjectedIntlProps;

class ContractOwner extends React.Component<Props, {}> {
  constructor(props: Props) {
    super(props);
    this.handleEmailClick = this.handleEmailClick.bind(this);
  }

  public render() {
    const { contractOwner } = this.props;
    const fullName = `${contractOwner.firstName} ${contractOwner.lastName}`;
    return (
      <div className="contract-owner" data-qa="contract-owner-section">
        <div className="contract-owner-questions">
          <FormattedMessage id="contractManager.questions" />
        </div>
        <h1>{fullName}</h1>
        <div className="contract-owner-label">
          <FormattedMessage id="contractManager.title" />
        </div>
        <div className="contract-owner-phone">
          {contractOwner.phone}
        </div>
        <a
          href={`mailto:${contractOwner.email}`}
          onClick={this.handleEmailClick}
        >
          <FormattedMessage id="contractManager.email" /> {fullName}
        </a>
      </div>
    );
  }

  private handleEmailClick() {
    this.props.emailContractOwner();
  }
}

const mapState = (state: State): SP => ({
  contractOwner: selectors.getContractOwner(state),
});

const mapDispatch = {
  emailContractOwner: actions.emailContractOwner,
};

export default connect<SP, DP, {}>(mapState, mapDispatch)(injectIntl(ContractOwner));
export { ContractOwner as PureContractOwner };
