// react-components-styleguide/styleguide.config.js
/* eslint-disable import/no-extraneous-dependencies */
const path = require('path');
const reactDocgen = require('react-docgen');
const reactDocgenTypescript = require('react-docgen-typescript');

const grcRoot = require.resolve('gca-react-components').replace('/src/index.js', '');

module.exports = {
  title: 'GCA React Components',
  serverPort: 3002,
  styleguideDir: 'dist',
  sections: [
    {
      name: 'Getting Started',
      content: path.join(grcRoot,'README.md'),
    },
    {
      name: 'Components',
      components: () => ([
        path.join(grcRoot,'src/components/AlertMessage/AlertMessage.js'),
        path.join(grcRoot,'src/components/Badge/Badge.tsx'),
        path.join(grcRoot,'src/components/BarChart/BarChart.js'),
        path.join(grcRoot,'src/components/Button/Button.js'),
        path.join(grcRoot,'src/components/ConfirmContent/ConfirmContent.js'),
        path.join(grcRoot,'src/components/Cropper/Cropper.js'),
        path.join(grcRoot,'src/components/Checkbox/Checkbox.js'),
        path.join(grcRoot,'src/components/DateInput/DateInput.js'),
        path.join(grcRoot,'src/components/DateRangePicker/DateRangePicker.js'),
        path.join(grcRoot,'src/components/ImageDropzone/ImageDropzone.js'),
        path.join(grcRoot,'src/components/Fieldset/Fieldset.js'),
        path.join(grcRoot,'src/components/Input/Input.js'),
        path.join(grcRoot,'src/components/Layout/Layout.js'),
        path.join(grcRoot,'src/components/LabelGroup/LabelGroup.js'),
        path.join(grcRoot,'src/components/LocaleSelect/LocaleSelect.js'),
        path.join(grcRoot,'src/components/Nav/Nav.js'),
        path.join(grcRoot,'src/components/Nav/NavIcon.js'),
        path.join(grcRoot,'src/components/Nav/NavLabel.js'),
        path.join(grcRoot,'src/components/PhoneInput/PhoneInput.js'),
        path.join(grcRoot,'src/components/PriceSlider/PriceSlider.js'),
        path.join(grcRoot,'src/components/ProfileHours/ProfileHours.js'),
        path.join(grcRoot,'src/components/RadioButton/RadioButton.js'),
        path.join(grcRoot,'src/components/RadioButton/SimpleRadioButtonGroup.js'),
        path.join(grcRoot,'src/components/RadioButton/RadioButtonGroup.js'),
        path.join(grcRoot,'src/components/Select/Select.js'),
        path.join(grcRoot,'src/components/StarRating/StarRating.js'),
        path.join(grcRoot,'src/components/SwitchList/SwitchList.js'),
        path.join(grcRoot,'src/components/TextArea/TextArea.js'),
        path.join(grcRoot,'src/components/TimeRange/TimeRange.tsx'),
        path.join(grcRoot,'src/components/TimeSelect/TimeSelect.js'),
      ]),
    },
    {
      name: 'redux-form fields',
      content: path.join(grcRoot,'src/reduxform-fields/ReduxFormFields.md'),
      components: () => ([
        path.join(grcRoot,'src/components/Checkbox/CheckboxField.js'),
        path.join(grcRoot,'src/components/Select/SelectField.js'),
        path.join(grcRoot,'src/components/TextArea/TextAreaField.js'),
        path.join(grcRoot,'src/components/TimeRange/TimeRangeField.js')
      ])
    },
    {
      name: 'final-form fields',
      content: path.join(grcRoot,'src/finalform-fields/FinalFormFields.md'),
      components: () => ([
        path.join(grcRoot,'src/finalform-fields/CheckboxFinalField.tsx'),
        path.join(grcRoot,'src/finalform-fields/DateInputFinalField.tsx'),
        path.join(grcRoot,'src/finalform-fields/DateRangePickerFinalField.tsx'),
        path.join(grcRoot,'src/finalform-fields/InputFinalField.tsx'),
        path.join(grcRoot,'src/finalform-fields/LocaleSelectFinalField.tsx'),
        path.join(grcRoot,'src/finalform-fields/PhoneInputFinalField.tsx'),
        path.join(grcRoot,'src/finalform-fields/PriceSliderFinalField.tsx'),
        path.join(grcRoot,'src/finalform-fields/ProfileHoursFinalField.tsx'),
        path.join(grcRoot,'src/finalform-fields/SelectFinalField.tsx'),
        path.join(grcRoot,'src/finalform-fields/SimpleRadioButtonGroupFinalField.tsx'),
        path.join(grcRoot,'src/finalform-fields/SwitchListFinalField.tsx'),
        path.join(grcRoot,'src/finalform-fields/TextAreaFinalField.tsx'),
        path.join(grcRoot,'src/finalform-fields/TimeRangeFinalField.tsx'),
        path.join(grcRoot,'src/finalform-fields/TimeSelectFinalField.tsx'),
      ])
    },
    {
      name: 'Icons',
      components: () => ([
        path.join(grcRoot,'src/icons/IconGallery.js'),
      ]),
    },
    {
      name: 'Redux Middleware',
      content: path.join(grcRoot,'docs/mixpanelMiddleware.md'),
    },
    {
      name: 'Contributing',
      content: path.join(grcRoot,'CONTRIBUTING.md'),
    },
  ],
  assetsDir: './static/',
  template: './index.html',
  theme: {
    fontFamily: {
      base: '"Source Sans Pro", sans-serif',
      monospace: '"Courier New", monospace'
    },
  },
  require: [
    'babel-polyfill',
    './styles/common.scss',
  ],
  getComponentPathLine(componentPath) {
    const name = path.basename(componentPath).split('.')[0];
    return `import { ${name} } from 'gca-react-components';`
  },
  propsParser: function(filePath, source, resolver, handlers) {
    if (/tsx?$/.test(filePath)) {
      return reactDocgenTypescript.parse(filePath);
    }
    return reactDocgen.parse(source, resolver, handlers);
  },
};
