// gc-microapp-venga/src/helpers/tracking.ts
import eventTracker from '@buffet/event-tracker';


const trackingCategory = 'Venga Onboarding';
eventTracker.queue('set', 'dimension4', trackingCategory);
eventTracker.enable();

// tslint:disable:object-literal-sort-keys

// Actions and Labels
const trackingEvents = {
  marketingPage: {
    action: 'marketing page',
    labels: {
      getVengaNow: 'get venga now',
      selectReviewManager: 'select review manager',
      selectHost: 'select host',
      compareOpen: 'compare open',
      compareClose: 'compare close',
      weCanHelp: 'we can help',
    },
  },
  signUpPage: {
    action: 'sign up page',
    labels: {
      back: 'back',
      checkboxChecked: 'checkbox - checked',
      checkboxUnchecked: 'checkbox - unchecked',
      showTerms: 'show terms',
      hideTerms: 'hide terms',
      sign: 'sign agreement',
      signComplete: 'sign agreement - complete',
    },
  },
  activePage: {
    action: 'active page',
    labels: {
      launchToVenga: 'launch to venga',
    },
  },
};

export { eventTracker, trackingCategory, trackingEvents };
