// damper/src/components/Area/Line.story.js
import React from 'react';
import { boolean, number, color, select, text } from '@storybook/addon-knobs';
import * as d3 from 'd3';
import { Area } from '../Area';
import theme from '../../theme/defaultTheme';

const { scaleLinear } = d3;

const graphHeight = 450;
const areaData = [
  {
    key: 'a',
    style: {
      val: 207,
      x: 0,
      y0: 410,
      y1: graphHeight,
    },
  },
  {
    key: 'b',
    style: {
      val: 29,
      x: 200,
      y0: 383,
      y1: graphHeight,
    },
  },
  {
    key: 'c',
    style: {
      val: 45,
      x: 400,
      y0: graphHeight,
      y1: graphHeight,
    },
  },
  {
    key: 'd',
    style: {
      val: 45,
      x: 600,
      y0: 32,
      y1: graphHeight,
    },
  },
  {
    key: 'e',
    style: {
      val: 45,
      x: 800,
      y0: 352,
      y1: graphHeight,
    },
  },
];

const activeInnerArea = {
  top: 0,
  height: graphHeight,
  scales: {
    x: scaleLinear().domain([0, 2]).range([30, (800 - 30)]),
    yInverted: scaleLinear().domain([0, 100]).range([graphHeight, 30]),
  },
};

class LineStory extends React.Component {
  render() {
    const lineOpacityOption = {
      range: true,
      min: 0,
      max: 1,
      step: 0.1,
    };
    const lineWidthOption = {
      range: true,
      min: 1,
      max: 20,
      step: 1,
    };

    const lineCurveOptions = [
      'curveMonotoneX',
      'curveLinear',
      'curveCardinal',
      'curveStep',
    ];

    const isMounted = boolean('isMounted', true);
    const isAnimated = boolean('isAnimated', true);
    const lineCurve = select('Curve', lineCurveOptions, lineCurveOptions[0]);
    const lineOpacity = number('Opacity', 1, lineOpacityOption);
    const lineWidth = number('Width', 1, lineWidthOption);
    const lineColor = color('Color', '#9D82ED');
    const lineDashArray = text('Dash array', '10, 5');

    let customisedTheme = theme;

    customisedTheme = {
      ...customisedTheme,
      area: {
        ...customisedTheme.area,
        curve: lineCurve,
      },
      common: {
        ...customisedTheme.common,
        fill: 'transparent',
        stroke: lineColor,
        strokeOpacity: lineOpacity,
        strokeWidth: lineWidth,
        strokeDasharray: lineDashArray,
        mouseOver: {
          ...customisedTheme.common.mouseOver,
          fill: 'transparent',
          stroke: lineColor,
          strokeOpacity: lineOpacity,
        },
      },
    };

    return (
      <article className="markdown-body">
        <h1>Line Component</h1>
        <svg width={800} height={graphHeight}>
          {isMounted && (
            <Area
              actions={{
                onClick: () => {},
                onMouseOver: () => {},
                onMouseOut: () => {},
              }}
              id="storybook-test"
              index={1}
              animation={{
                animate: isAnimated,
              }}
              scale={(d) => d}
              theme={customisedTheme}
              data={areaData}
              activeInnerArea={activeInnerArea}
            />
          )}
        </svg>
        <hr />
        <p>
        Current settings for the previewed line are;<br />
          isAnimated: {`${isAnimated}`}<br />
          Curve: {`${lineCurve}`}<br />
          Opacity: {`${lineOpacity}`}<br />
          Width: {`${lineWidth}`}<br />
          Color: {`${lineColor}`}<br />
          Dash array: {`${lineDashArray}`}<br />
        </p>
        <p>
          Note: if a fill is transparent and a stroke is defined in the theme , a line graph is generated instead of an area.
        </p>
      </article>
    );
  }
}

export { LineStory };
