// gca-react-components/src/redux-modules/languageProvider.js
import safeGet from '../utils/safeGet';

/**
 *
 * LanguageProvider constants
 *
 */

const CHANGE_LOCALE = 'gca/languageProvider/CHANGE_LOCALE';

/**
 *
 * LanguageProvider actions
 *
 */

export function changeLocale(languageLocale) {
  return {
    type: CHANGE_LOCALE,
    locale: languageLocale,
  };
}

/**
 *
 * LanguageProvider selectors
 *
 */

export const selectors = {
  locale(state) {
    return safeGet(state, ['language', 'locale']);
  },
};

/**
 *
 * LanguageProvider reducer & factory
 *
 */

export const defaultInitialState = {
  locale: 'en-us',
};

/**
 * @param {object} initialState - the initial state to be set as a default argument in the reducer
 * @return {function} reducer - the final reducer function
 */
export function createReducer(initialState = defaultInitialState) {
  return function reducer(state = initialState, action = {}) {
    switch (action.type) {
      case CHANGE_LOCALE:
        return { ...state, locale: action.locale };
      default:
        return state;
    }
  };
}

export const languageProviderReducer = createReducer(defaultInitialState);
