// user-agreements/src/TermsAndConditions/TermsCallToAction.tsx
import React from 'react';
import { FormattedMessage, InjectedIntlProps, injectIntl } from 'react-intl';
import './TermsCallToAction.scss';

export const ACCEPT_CTA_STATE = 'acceptCTAState';
export const SCROLL_CTA_STATE = 'scrollCTAState';

interface OP {
  ctaState: string
}

type Props = OP & InjectedIntlProps;

class TermsCallToAction extends React.Component<Props, {}> {
  public render() {
    let scrollClasses = 'scroll-cta';
    let acceptClasses = 'accept-cta';

    if (this.props.ctaState === ACCEPT_CTA_STATE) {
      acceptClasses += ' show';
    } else if (this.props.ctaState === SCROLL_CTA_STATE) {
      scrollClasses += ' show';
    }

    return (
      <>
        <div className={scrollClasses} data-qa="terms-scroll-cta">
          <FormattedMessage id="termsAndConditions.scrollCTA" />
        </div>
        <div className={acceptClasses} data-qa="terms-accept-cta">
          <FormattedMessage id="termsAndConditions.acceptCTA" />
        </div>
      </>
    );
  }
}

export default injectIntl(TermsCallToAction);
export { TermsCallToAction as PureTermsCallToAction };
