// user-accounts/src/store/reducers.ts
/**
 * Combine all reducers in this file and export the combined reducers.
 * If we were to do this in store.js, reducers wouldn't be hot reloadable.
 */
import { currentContextReducer } from 'gca-react-components/src/redux-modules/currentContext';
import { featureTogglesReducer } from 'gca-react-components/src/redux-modules/featureToggles';
import { createReducer as createLanguageReducer } from 'gca-react-components/src/redux-modules/languageProvider';
import get from 'lodash/get';
import { combineReducers } from 'redux';

import { reducer as availableUserReducer } from './redux-modules/availableUser';
import { reducer as currentUserReducer } from './redux-modules/currentUser';
import { modalReducer } from './redux-modules/modals';
import { reducer as rolesReducer } from './redux-modules/roles';
import { reducer as scopeReducer } from './redux-modules/scope';
import { reducer as usersReducer } from './redux-modules/users';
import { State } from './types';

// Retrieve the locale code from the restaurant gcaConfig or the group groupSettingsConfig
const appType = get(window, ['gcScope', 'type']);
const config = appType === 'restaurant' ? 'gcaConfig' : 'groupSettingsConfig';
const locale = get(window, [config, 'i18n', 'primaryLocale']);
/**
 * Creates the main reducer from all other reducers.
 */
export default function createReducer(initialState: Partial<State> | undefined) {
  return combineReducers<State>({
    availableUser: availableUserReducer,
    currentContext: currentContextReducer,
    currentUser: currentUserReducer,
    // feature toggles are not required for group
    featureToggles: get(initialState, ['scope', 'type']) === 'group' ? () => ({}) : featureTogglesReducer,
    language: createLanguageReducer({ locale: locale || 'en-us' }),
    modals: modalReducer,
    // tslint:disable-next-line
    appRoles: rolesReducer('appRoles'),
    profileRoles: rolesReducer('profileRoles'),
    // tslint:disable-next-line
    appScope: appScope => (appScope || {}),
    scope: scopeReducer,
    users: usersReducer,
  });
}
