// gca-react-components/src/components/ModalWrapper/ModalWrapper.js
import React from 'react';
import PropTypes from 'prop-types';

import { Button } from '../Button/Button';
import Modal from '../Modal/Modal';

import './ModalWrapper.scss';

export const ModalWrapper = props => {
  const { onConfirm, title, onCancel, confirmButtonText, cancelButtonText, includeCancelButton } = props;

  const cancelButton = includeCancelButton ? (
    <Button className="gca-modal-button" onClick={onCancel}>
      {cancelButtonText}
    </Button>
  ) : null;

  /* eslint-disable jsx-a11y/no-static-element-interactions */
  return (
    <Modal onClose={onCancel} overlayClass="gca-modal-wrapper-overlay" modalClass="gca-modal-wrapper">
      <header>
        <h1>{title}</h1>
      </header>
      <main>{props.children}</main>
      <footer>
        {cancelButton}
        <Button theme="primary" className="gca-modal-button" onClick={onConfirm}>
          {confirmButtonText}
        </Button>
      </footer>
    </Modal>
  );
};

ModalWrapper.propTypes = {
  title: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired,
  onConfirm: PropTypes.func.isRequired,
  confirmButtonText: PropTypes.string.isRequired,
  onCancel: PropTypes.func.isRequired,
  cancelButtonText: PropTypes.string,
  includeCancelButton: PropTypes.bool,
};

ModalWrapper.defaultProps = {
  includeCancel: false,
};
