// @buffet/scope-selector/src/service/RestaurantMapper.ts
import { Restaurant, RestaurantDTO } from './ScopeSelector.types';

export function mapRestaurants(restaurants: RestaurantDTO[]): Restaurant[] {
  return restaurants.map(normalizeRestaurant);
}

/**
 * There are three possible formats for restaurant objects
 * 1. Results coming from restaurant search have an `addressResponse` property
 * 2. Results coming from context or switch RID endpoints have an `address` property.
 * 3. The normalized objects in localStorage have `city` and `country` properties.
 *
 * This function converts 1 & 2 into 3.
 */
export function normalizeRestaurant(restaurant: RestaurantDTO): Restaurant {
  return {
    rid: restaurant.rid,
    name: restaurant.name,
    isTestRestaurant: restaurant.testRestaurant || restaurant.isTestRestaurant || false,
    city: getCityFromDTO(restaurant),
    country: getCountryFromDTO(restaurant)
  };
}

function getCityFromDTO(restaurant: RestaurantDTO): string | undefined {
  if (restaurant.city) {
    return restaurant.city;
  }

  if (restaurant.addressResponse && restaurant.addressResponse.city) {
    return restaurant.addressResponse.city;
  }

  if (restaurant.address && restaurant.address.city) {
    return restaurant.address.city;
  }

  return undefined;
}

function getCountryFromDTO(restaurant: RestaurantDTO): string | undefined {
  if (restaurant.country) {
    return restaurant.country;
  }

  if (restaurant.addressResponse && restaurant.addressResponse.country) {
    return restaurant.addressResponse.country;
  }

  if (restaurant.address && restaurant.address.countryCode) {
    return restaurant.address.countryCode;
  }

  return undefined;
}
