// gca-react-components/src/components/IntroVideoModal/IntroVideoModal.js
import React from 'react';
import PropTypes from 'prop-types';
import { injectIntl, intlShape, FormattedMessage } from 'react-intl';

import { ModalWrapper } from '../../components/ModalWrapper/ModalWrapper';
import './IntroVideoModal.scss';

export const PureIntroVideoModal = props => (
  <ModalWrapper
    title={props.intl.formatMessage({ id: props.titleId })}
    confirmButtonText={props.intl.formatMessage({ id: props.buttonId })}
    onConfirm={props.onClose}
    onCancel={props.onClose}
  >
    <div className="gca-intro-video-body">
      <FormattedMessage tagName="p" id={props.bodyId} />
      <iframe title="Intro Video" width="560" height="315" src={props.videoSource} frameBorder="0" allowFullScreen />
    </div>
  </ModalWrapper>
);

PureIntroVideoModal.propTypes = {
  onClose: PropTypes.func.isRequired,
  titleId: PropTypes.string.isRequired,
  buttonId: PropTypes.string.isRequired,
  bodyId: PropTypes.string.isRequired,
  videoSource: PropTypes.string.isRequired,
  intl: intlShape,
};

export const IntroVideoModal = injectIntl(PureIntroVideoModal);
