// @buffet/color-tokens/dist/tokens.common.js

module.exports = {
  grayNavBorder: "rgb(221, 221, 221)",
  grayDisabledButtonText: "rgb(180, 182, 191)",
  grayPanelBackground: "rgb(240, 241, 244)",
  grayFootText: "rgb(179, 179, 179)",
  grayPageBackground: "rgb(240, 240, 245)",
  grayTextSubtle: "rgb(146, 148, 149)",
  grayNavFocus: "rgb(58, 59, 66)",
  grayEmptyTable: "rgb(118, 122, 132)",
  grayTableTitles: "rgb(152, 152, 154)",
  greenPaidChairs: "rgb(112, 158, 54)",
  purpleSeatedChairs: "rgb(131, 98, 163)",
  grayStandard: "rgb(160, 168, 189)",
  purplePartiallySeatedChairs: "rgb(120, 103, 172)",
  grayGutterOff: "rgb(34, 36, 40)",
  graySearchLighter: "rgb(206, 209, 217)",
  greenDropdownTag: "rgb(65, 185, 138)",
  white: "rgb(255, 255, 255)",
  grayHintText: "rgb(162, 168, 183)",
  greenCourse1: "rgb(33, 141, 175)",
  greenDroppedCheckChairs: "rgb(68, 151, 94)",
  grayStatusIcon: "rgb(162, 168, 183)",
  grayDropdownHover: "rgb(48, 51, 55)",
  greenCourse2: "rgb(57, 174, 197)",
  grayTagText: "rgb(112, 117, 129)",
  grayTableHighlight: "rgb(247, 247, 247)",
  grayButtonHover: "rgb(240, 240, 240)",
  grayFloorplan: "rgb(72, 76, 83)",
  greenCourse3: "rgb(84, 186, 186)",
  graySearchLightest: "rgb(240, 241, 245)",
  redError: "rgb(244, 78, 64)",
  grayBusyLight: "rgb(230, 231, 235)",
  greenCourse4: "rgb(132, 197, 184)",
  blueLinkText: "rgb(79, 152, 219)",
  blueCancelText: "rgb(0, 0, 0)",
  blueSeatButtonHighlight: "rgb(199, 223, 243)",
  grayDropdownDisabled: "rgb(107, 113, 124)",
  grayNavPress: "rgb(46, 47, 52)",
  greenCourse4Chairs: "rgb(87, 153, 139)",
  grayBodyBackground: "rgb(35, 36, 38)",
  bluePrimary: "rgb(108, 182, 245)",
  greenCourse3Chairs: "rgb(67, 145, 145)",
  graySearchDark: "rgb(123, 128, 140)",
  greenCourse2Chairs: "rgb(44, 137, 156)",
  greenDroppedCheck: "rgb(84, 192, 119)",
  grayPlaceholderText: "rgb(181, 182, 184)",
  grayHeadBackground: "rgb(70, 73, 80)",
  grayDividerLine: "rgb(238, 238, 238)",
  greenCourse1Chairs: "rgb(53, 94, 110)",
  grayEmptyTableNumber: "rgb(72, 76, 83)",
  greenGo: "rgb(142, 196, 74)",
  grayOutlinedButton: "rgb(199, 203, 212)",
  graySeatButtonShaded: "rgb(226, 227, 232)",
  purplePartiallySeated: "rgb(147, 127, 214)",
  purpleSeated: "rgb(164, 122, 206)",
  grayErrorPanel: "rgb(235, 236, 238)",
  grayTableBorder: "rgb(213, 216, 222)",
  graySecondaryButton: "rgb(150, 158, 181)",
  grayTableHeader: "rgb(250, 250, 250)",
  blueHighlightAccent: "rgb(74, 130, 183)",
  grayTableText: "rgb(49, 50, 53)",
  greenConfirm: "rgb(154, 204, 84)",
  grayTableStripe: "rgb(249, 249, 249)",
  grayHeaderShadow: "rgb(28, 29, 33)",
  greenPaid: "rgb(142, 196, 74)",
  grayHeader: "rgb(51, 53, 58)",
  grayText: "rgb(30, 32, 34)",
  graySeatButtonHighlight: "rgb(199, 223, 243)",
  grayDropdownActive: "rgb(36, 38, 42)",
  grayBackground: "rgb(58, 60, 66)",
  grayDisabledButton: "rgb(215, 217, 224)",
  grayPanel: "rgb(238, 239, 243)",
  grayDropdown: "rgb(60, 63, 69)",
  grayEmptyChair: "rgb(92, 96, 104)",
  graySearchLight: "rgb(172, 177, 191)",
  blueSeatButtonShaded: "rgb(226, 227, 232)",
  orangeWarning: "rgb(244, 156, 27)",
  graySearchDarker: "rgb(57, 59, 60)",
  grayListPress: "rgb(234, 234, 234)",
  redDanger: "rgb(245, 89, 73)",
  grayPanelBorder: "rgb(214, 217, 222)",
  grayDropdownText: "rgb(153, 153, 153)",
  black: "rgb(0, 0, 0)",
};
