// @buffet/color-tokens/dist/tokens.module.js

export const grayNavBorder = "rgb(221, 221, 221)";
export const grayDisabledButtonText = "rgb(180, 182, 191)";
export const grayPanelBackground = "rgb(240, 241, 244)";
export const grayFootText = "rgb(179, 179, 179)";
export const grayPageBackground = "rgb(240, 240, 245)";
export const grayTextSubtle = "rgb(146, 148, 149)";
export const grayNavFocus = "rgb(58, 59, 66)";
export const grayEmptyTable = "rgb(118, 122, 132)";
export const grayTableTitles = "rgb(152, 152, 154)";
export const greenPaidChairs = "rgb(112, 158, 54)";
export const purpleSeatedChairs = "rgb(131, 98, 163)";
export const grayStandard = "rgb(160, 168, 189)";
export const purplePartiallySeatedChairs = "rgb(120, 103, 172)";
export const grayGutterOff = "rgb(34, 36, 40)";
export const graySearchLighter = "rgb(206, 209, 217)";
export const greenDropdownTag = "rgb(65, 185, 138)";
export const white = "rgb(255, 255, 255)";
export const grayHintText = "rgb(162, 168, 183)";
export const greenCourse1 = "rgb(33, 141, 175)";
export const greenDroppedCheckChairs = "rgb(68, 151, 94)";
export const grayStatusIcon = "rgb(162, 168, 183)";
export const grayDropdownHover = "rgb(48, 51, 55)";
export const greenCourse2 = "rgb(57, 174, 197)";
export const grayTagText = "rgb(112, 117, 129)";
export const grayTableHighlight = "rgb(247, 247, 247)";
export const grayButtonHover = "rgb(240, 240, 240)";
export const grayFloorplan = "rgb(72, 76, 83)";
export const greenCourse3 = "rgb(84, 186, 186)";
export const graySearchLightest = "rgb(240, 241, 245)";
export const redError = "rgb(244, 78, 64)";
export const grayBusyLight = "rgb(230, 231, 235)";
export const greenCourse4 = "rgb(132, 197, 184)";
export const blueLinkText = "rgb(79, 152, 219)";
export const blueCancelText = "rgb(0, 0, 0)";
export const blueSeatButtonHighlight = "rgb(199, 223, 243)";
export const grayDropdownDisabled = "rgb(107, 113, 124)";
export const grayNavPress = "rgb(46, 47, 52)";
export const greenCourse4Chairs = "rgb(87, 153, 139)";
export const grayBodyBackground = "rgb(35, 36, 38)";
export const bluePrimary = "rgb(108, 182, 245)";
export const greenCourse3Chairs = "rgb(67, 145, 145)";
export const graySearchDark = "rgb(123, 128, 140)";
export const greenCourse2Chairs = "rgb(44, 137, 156)";
export const greenDroppedCheck = "rgb(84, 192, 119)";
export const grayPlaceholderText = "rgb(181, 182, 184)";
export const grayHeadBackground = "rgb(70, 73, 80)";
export const grayDividerLine = "rgb(238, 238, 238)";
export const greenCourse1Chairs = "rgb(53, 94, 110)";
export const grayEmptyTableNumber = "rgb(72, 76, 83)";
export const greenGo = "rgb(142, 196, 74)";
export const grayOutlinedButton = "rgb(199, 203, 212)";
export const graySeatButtonShaded = "rgb(226, 227, 232)";
export const purplePartiallySeated = "rgb(147, 127, 214)";
export const purpleSeated = "rgb(164, 122, 206)";
export const grayErrorPanel = "rgb(235, 236, 238)";
export const grayTableBorder = "rgb(213, 216, 222)";
export const graySecondaryButton = "rgb(150, 158, 181)";
export const grayTableHeader = "rgb(250, 250, 250)";
export const blueHighlightAccent = "rgb(74, 130, 183)";
export const grayTableText = "rgb(49, 50, 53)";
export const greenConfirm = "rgb(154, 204, 84)";
export const grayTableStripe = "rgb(249, 249, 249)";
export const grayHeaderShadow = "rgb(28, 29, 33)";
export const greenPaid = "rgb(142, 196, 74)";
export const grayHeader = "rgb(51, 53, 58)";
export const grayText = "rgb(30, 32, 34)";
export const graySeatButtonHighlight = "rgb(199, 223, 243)";
export const grayDropdownActive = "rgb(36, 38, 42)";
export const grayBackground = "rgb(58, 60, 66)";
export const grayDisabledButton = "rgb(215, 217, 224)";
export const grayPanel = "rgb(238, 239, 243)";
export const grayDropdown = "rgb(60, 63, 69)";
export const grayEmptyChair = "rgb(92, 96, 104)";
export const graySearchLight = "rgb(172, 177, 191)";
export const blueSeatButtonShaded = "rgb(226, 227, 232)";
export const orangeWarning = "rgb(244, 156, 27)";
export const graySearchDarker = "rgb(57, 59, 60)";
export const grayListPress = "rgb(234, 234, 234)";
export const redDanger = "rgb(245, 89, 73)";
export const grayPanelBorder = "rgb(214, 217, 222)";
export const grayDropdownText = "rgb(153, 153, 153)";
export const black = "rgb(0, 0, 0)";
