// gca-react-components/styleguide.config.js
const path = require('path');
const reactDocgen = require('react-docgen');
const reactDocgenTypescript = require('react-docgen-typescript');

module.exports = {
  title: 'GCA React Components',
  serverPort: 3002,
  sections: [
    {
      name: 'Getting Started',
      content: './README.md',
    },
    {
      name: 'Components',
      components: () => ([
        './src/components/AlertMessage/AlertMessage.js',
        './src/components/Badge/Badge.js',
        './src/components/BarChart/BarChart.js',
        './src/components/Button/Button.js',
        './src/components/ConfirmContent/ConfirmContent.js',
        './src/components/Cropper/Cropper.js',
        './src/components/Checkbox/Checkbox.js',
        './src/components/DateInput/DateInput.js',
        './src/components/DateRangePicker/DateRangePicker.js',
        './src/components/ImageDropzone/ImageDropzone.js',
        './src/components/Fieldset/Fieldset.js',
        './src/components/Input/Input.js',
        './src/components/Layout/Layout.js',
        './src/components/LabelGroup/LabelGroup.js',
        './src/components/LocaleSelect/LocaleSelect.js',
        './src/components/Nav/Nav.js',
        './src/components/Nav/NavIcon.js',
        './src/components/Nav/NavLabel.js',
        './src/components/PhoneInput/PhoneInput.js',
        './src/components/PriceSlider/PriceSlider.js',
        './src/components/ProfileHours/ProfileHours.js',
        './src/components/RadioButton/RadioButton.js',
        './src/components/RadioButton/SimpleRadioButtonGroup.js',
        './src/components/RadioButton/RadioButtonGroup.js',
        './src/components/Select/Select.js',
        './src/components/StarRating/StarRating.js',
        './src/components/SwitchList/SwitchList.js',
        './src/components/TextArea/TextArea.js',
        './src/components/TimeRange/TimeRange.tsx',
        './src/components/TimeSelect/TimeSelect.js',
      ]),
    },
    {
      name: 'redux-form fields',
      content: './src/reduxform-fields/ReduxFormFields.md',
      components: () => ([
        './src/components/Checkbox/CheckboxField.js',
        './src/components/Select/SelectField.js',
        './src/components/TextArea/TextAreaField.js',
        './src/components/TimeRange/TimeRangeField.js'
      ])
    },
    {
      name: 'final-form fields',
      content: './src/finalform-fields/FinalFormFields.md',
      components: () => ([
        './src/finalform-fields/CheckboxFinalField.tsx',
        './src/finalform-fields/DateInputFinalField.tsx',
        './src/finalform-fields/DateRangePickerFinalField.tsx',
        './src/finalform-fields/InputFinalField.tsx',
        './src/finalform-fields/LocaleSelectFinalField.tsx',
        './src/finalform-fields/PhoneInputFinalField.tsx',
        './src/finalform-fields/PriceSliderFinalField.tsx',
        './src/finalform-fields/ProfileHoursFinalField.tsx',
        './src/finalform-fields/SelectFinalField.tsx',
        './src/finalform-fields/SimpleRadioButtonGroupFinalField.tsx',
        './src/finalform-fields/SwitchListFinalField.tsx',
        './src/finalform-fields/TextAreaFinalField.tsx',
        './src/finalform-fields/TimeRangeFinalField.tsx',
        './src/finalform-fields/TimeSelectFinalField.tsx',
      ])
    },
    {
      name: 'Utilities',
      components: () => ([
        './src/utils/validate.ts',
      ]),
    },
    {
      name: 'Icons',
      components: () => ([
        './src/icons/IconGallery.js',
      ]),
    },
    {
      name: 'Redux Middleware',
      content: './docs/mixpanelMiddleware.md',
    },
    {
      name: 'Contributing',
      content: './CONTRIBUTING.md',
    },
    {
      name: 'Change Log',
      content: './CHANGELOG.md',
    },
  ],
  assetsDir: './internals/static/',
  template: 'src/index.html',
  theme: {
    fontFamily: {
      base: '"Source Sans Pro", sans-serif',
      monospace: '"Courier New", monospace'
    },
  },
  require: [
    'babel-polyfill',
    path.join(__dirname, 'src/styles/common.scss'),
  ],
  propsParser(filePath, source) {
    if (/tsx?$/.test(filePath)) {
      return reactDocgenTypescript.parse(filePath);
    }

    return reactDocgen.parse(source);
  },
};
