// final-form-arrays/dist/final-form-arrays.es.js
//      
var insert = function insert(_ref, state, _ref2) {
  var name = _ref[0],
      index = _ref[1],
      value = _ref[2];
  var changeValue = _ref2.changeValue;
  changeValue(state, name, function (array) {
    var copy = [].concat(array || []);
    copy.splice(index, 0, value);
    return copy;
  });
};

function _extends() {
  _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  return _extends.apply(this, arguments);
}

//      
var move = function move(_ref, state, _ref2) {
  var name = _ref[0],
      from = _ref[1],
      to = _ref[2];
  var changeValue = _ref2.changeValue;

  if (from === to) {
    return;
  }

  changeValue(state, name, function (array) {
    var copy = [].concat(array || []);
    var value = copy[from];
    copy.splice(from, 1);
    copy.splice(to, 0, value);
    return copy;
  });
  var fromPrefix = name + "[" + from + "]";
  Object.keys(state.fields).forEach(function (key) {
    if (key.substring(0, fromPrefix.length) === fromPrefix) {
      var suffix = key.substring(fromPrefix.length);
      var fromKey = fromPrefix + suffix;
      var backup = state.fields[fromKey];

      if (from < to) {
        // moving to a higher index
        // decrement all indices between from and to
        for (var i = from; i < to; i++) {
          var destKey = name + "[" + i + "]" + suffix;
          moveFieldState({
            destKey: destKey,
            source: state.fields[name + "[" + (i + 1) + "]" + suffix]
          });
        }
      } else {
        // moving to a lower index
        // increment all indices between to and from
        for (var _i = from; _i > to; _i--) {
          var _destKey = name + "[" + _i + "]" + suffix;

          moveFieldState({
            destKey: _destKey,
            source: state.fields[name + "[" + (_i - 1) + "]" + suffix]
          });
        }
      }

      var toKey = name + "[" + to + "]" + suffix;
      moveFieldState({
        destKey: toKey,
        source: backup
      });
    }
  });

  function moveFieldState(_ref3) {
    var destKey = _ref3.destKey,
        source = _ref3.source;
    state.fields[destKey] = _extends({}, source, {
      name: destKey,
      change: state.fields[destKey].change,
      // prevent functions from being overwritten
      blur: state.fields[destKey].blur,
      focus: state.fields[destKey].focus,
      lastFieldState: undefined // clearing lastFieldState forces renotification

    });
  }
};

//      
var pop = function pop(_ref, state, _ref2) {
  var name = _ref[0];
  var changeValue = _ref2.changeValue;
  var result;
  changeValue(state, name, function (array) {
    if (array) {
      if (!array.length) {
        return [];
      }

      result = array[array.length - 1];
      return array.slice(0, array.length - 1);
    }
  });
  return result;
};

//      
var push = function push(_ref, state, _ref2) {
  var name = _ref[0],
      value = _ref[1];
  var changeValue = _ref2.changeValue;
  changeValue(state, name, function (array) {
    return array ? [].concat(array, [value]) : [value];
  });
};

//      
var remove = function remove(_ref, state, _ref2) {
  var name = _ref[0],
      index = _ref[1];
  var changeValue = _ref2.changeValue;
  var returnValue;
  changeValue(state, name, function (array) {
    var copy = [].concat(array || []);
    returnValue = copy[index];
    copy.splice(index, 1);
    return copy;
  });
  return returnValue;
};

//      
var shift = function shift(_ref, state, _ref2) {
  var name = _ref[0];
  var changeValue = _ref2.changeValue;
  var result;
  changeValue(state, name, function (array) {
    if (array) {
      if (!array.length) {
        return [];
      }

      result = array[0];
      return array.slice(1, array.length);
    }
  });
  return result;
};

//      
var swap = function swap(_ref, state, _ref2) {
  var name = _ref[0],
      indexA = _ref[1],
      indexB = _ref[2];
  var changeValue = _ref2.changeValue;

  if (indexA === indexB) {
    return;
  }

  changeValue(state, name, function (array) {
    var copy = [].concat(array || []);
    var a = copy[indexA];
    copy[indexA] = copy[indexB];
    copy[indexB] = a;
    return copy;
  }); // swap all field state that begin with "name[indexA]" with that under "name[indexB]"

  var aPrefix = name + "[" + indexA + "]";
  var bPrefix = name + "[" + indexB + "]";
  Object.keys(state.fields).forEach(function (key) {
    if (key.substring(0, aPrefix.length) === aPrefix) {
      var suffix = key.substring(aPrefix.length);
      var aKey = aPrefix + suffix;
      var bKey = bPrefix + suffix;
      var fieldA = state.fields[aKey];
      moveFieldState({
        destKey: aKey,
        source: state.fields[bKey]
      });
      moveFieldState({
        destKey: bKey,
        source: fieldA
      });
    }
  });

  function moveFieldState(_ref3) {
    var destKey = _ref3.destKey,
        source = _ref3.source;
    state.fields[destKey] = _extends({}, source, {
      name: destKey,
      change: state.fields[destKey].change,
      // prevent functions from being overwritten
      blur: state.fields[destKey].blur,
      focus: state.fields[destKey].focus,
      lastFieldState: undefined // clearing lastFieldState forces renotification

    });
  }
};

//      
var unshift = function unshift(_ref, state, _ref2) {
  var name = _ref[0],
      value = _ref[1];
  var changeValue = _ref2.changeValue;
  changeValue(state, name, function (array) {
    return array ? [value].concat(array) : [value];
  });
};

//      
var update = function update(_ref, state, _ref2) {
  var name = _ref[0],
      index = _ref[1],
      value = _ref[2];
  var changeValue = _ref2.changeValue;
  changeValue(state, name, function (array) {
    var copy = [].concat(array || []);
    copy.splice(index, 1, value);
    return copy;
  });
};

//      
var mutators = {
  insert: insert,
  move: move,
  pop: pop,
  push: push,
  remove: remove,
  shift: shift,
  swap: swap,
  unshift: unshift,
  update: update
};

export default mutators;
