// gc-microapp-pos/src/PosOnboarding/OnboardingDisplayState.tsx
import React from 'react';
import { OnboardingDisplayStateName } from '../store/types';

// other pages states to display
import LandingPage from '../LandingPage/LandingPage';
import NextSteps from '../NextStepsPage/NextStepsPage';
import PosSelectPage from '../PosSelectPage/PosSelectPage';
import PosStatusPage from '../PosStatusPage/PosStatusPage';
import SubmittedLeadGenPage from '../SubmittedLeadGenPage/SubmittedLeadGenPage';
import UnsupportedPosPage from '../UnsupportedPosPage/UnsupportedPosPage';

export interface Props {
  displayKey: OnboardingDisplayStateName;
}

type StatesToDisplay = { [key in OnboardingDisplayStateName]: React.ComponentType };

class OnboardingDisplayState extends React.PureComponent<Props> {
  private statesToDisplay: StatesToDisplay = {
    active: PosStatusPage,
    landing: LandingPage,
    nextSteps: NextSteps,
    posSelect: PosSelectPage,
    submittedLeadGen: SubmittedLeadGenPage,
    unsupportedPos: UnsupportedPosPage,
  };

  public render() {
    const CurrentStateToDisplay = this.statesToDisplay[this.props.displayKey] || this.statesToDisplay.landing;
    return <CurrentStateToDisplay />;
  }
}

export default OnboardingDisplayState;
