// gc-microapp-marketing/src/PromotedInventory/Insights/common/Tooltip/Tooltip.tsx
import React from 'react';
import HoverCard from '../HoverCard/HoverCard';
import SpeechDialog from '../SpeechDialog/SpeechDialog';

interface Props {
  tooltipProps: any;
  direction: string;
  location: {
    x: number;
    y: number;
  };
  dimensions: {
    width: string;
    height: string;
  }
}

// tslint:disable:object-literal-sort-keys
class Tooltip extends React.Component<Props> {
  public getHoverCardProps() {
    const { title, description, content } = this.props.tooltipProps;
    return {
      title,
      description,
      content,
    };
  }

  public getDivStyle() {
    const { location, dimensions } = this.props;
    return {
      position: 'absolute',
      width: dimensions.width,
      height: dimensions.height,
      left: location.x,
      top: location.y,
      pointerEvents: 'none',
    };
  }

  public render() {
    const { direction } = this.props;
    return (
      <div style={this.getDivStyle() as any /* TODO */}>
        <SpeechDialog direction={direction}>
          <div>
            <HoverCard {...this.getHoverCardProps()} />
          </div>
        </SpeechDialog>
      </div>
    );
  }
}

export default Tooltip;
