// gc-microapp-pos/src/PosSelectPage/PosSelectionTile.tsx
import React from 'react';
import { SupportedPosBrand, SupportedPosBrandsIds } from '../store/types';

import alohaLogo from '../assets/posLogos/aloha-large.png';
import dinerwareLogo from '../assets/posLogos/dinerware-large.png';
import hnlLogo from '../assets/posLogos/hnl.png';
import imposLogo from '../assets/posLogos/impos.png';
import microsLogo from '../assets/posLogos/micros-large.png';
import simphonyLogo from '../assets/posLogos/micros-simphony-large.png';
import positouchLogo from '../assets/posLogos/positouch-large.png';
import smsLogo from '../assets/posLogos/sms-large.png';
import squirrelLogo from '../assets/posLogos/squirrel-large.png';
import toastLogo from '../assets/posLogos/toast-large.png';
import upserveLogo from '../assets/posLogos/upserve-large.png';

import './PosSelectionTile.scss';

interface Props {
  posBrand: SupportedPosBrand;
  isSelected: boolean;
  selectPosAction: (payload: SupportedPosBrand) => void;
}

type PosLogoMap = { [posName in SupportedPosBrandsIds]: any }; // TODO: figure out type checking for png later

class PosSelectionTile extends React.PureComponent<Props> {
  private posLogoMap: PosLogoMap = {
    a4y0c000000LZ0fAAG: smsLogo,
    a4y0c000000UphzAAC: alohaLogo,
    a4y0c000000Upi0AAC: toastLogo,
    a4y0c000000Upi1AAC: microsLogo,
    a4y0c000000Upi2AAC: positouchLogo,
    a4y0c000000Upi3AAC: squirrelLogo,
    a4y0c000000Upi4AAC: dinerwareLogo,
    a4y0c000000Upi5AAC: simphonyLogo,
    a4y0c000000Upi7AAC: hnlLogo,
    a4y0c000000Upi8AAC: imposLogo,
    a4y0c000000UpiDAAS: upserveLogo,
    [SupportedPosBrandsIds.unselected]: null,
  };
  public render() {
    const { posBrand, isSelected, selectPosAction } = this.props;
    return (
      <div className={`pos-select__tile ${isSelected ? 'selected' : ''}`}>
        <input
          className="pos-select__tile__input"
          type="radio"
          id={posBrand.Id}
          value={posBrand.Id}
          name="posSelect"
          checked={isSelected}
          onChange={() => selectPosAction(posBrand)}
        />
        <label className="pos-select__tile__label" htmlFor={posBrand.Id}>
          <img className="pos-select__tile__logo" src={this.posLogoMap[posBrand.Id]} />
        </label>
      </div>
    );
  }
}

export default PosSelectionTile;
export { Props };
