// user-agreements/src/AgreementFooter/AgreementFooter.tsx
import { PublishButton } from 'gca-react-components';
import React from 'react';

import './AgreementFooter.scss';

interface OP {
  buttonText: string;
  disabled?: boolean;
  busy?: boolean;
  onButtonClick: () => void;
}

type Props = OP;

class AgreementFooter extends React.Component<Props, {}> {

  public state = {
    isShown: false,
  }

  public componentDidMount() {

    //
    // :TRICKY:
    // timeout state change in order to add 'show' class to footer after rendering so that the footer css transition
    // to slide up is triggered.
    //
    window.setTimeout(() => this.setState({ isShown: true }), 250);
  }

  public render() {
    const footerClasses =  this.state.isShown ? 'page-footer show' : 'page-footer';
    return (
      <div className={footerClasses} data-qa="agreement-footer">
        <div className="page-footer-content">
          <div className="button-container">
            <PublishButton
              buttonText={this.props.buttonText}
              disabled={this.props.disabled}
              busy={this.props.busy}
              successText={this.props.buttonText}
              onClick={this.handlePublishClick}
            />
          </div>
        </div>
      </div>
    )
  }

  private handlePublishClick = () => {
    this.props.onButtonClick();
  }
}

export default AgreementFooter;
