// gc-microapp-marketing/src/Gifts/TermsAndConditions/TermsAndConditions.tsx
import { Button, Checkbox } from 'gca-react-components';
import { CheckboxFinalField, InputFinalField } from 'gca-react-components/src/finalform-fields';
import { selectors, User } from 'gca-react-components/src/redux-modules/currentContext';
import React, { Component, SFC } from 'react';
import { Form } from 'react-final-form';
import { FormattedMessage } from 'react-intl';
import { connect } from 'react-redux';
import { actions } from '../../store/gifts/gifts';
import { ClickOrInteractPayload } from '../../store/gifts/gifts.types';
import { State } from '../../store/types';
import { LINK_TO_GIFTS_TERMS_V1 } from '../constants';
import FinalFieldFocusTracker from '../FinalFieldFocusTracker';
import validateGiftsTermsForm from './validate';

import './TermsAndConditions.scss';

interface TermsFormValues {
  name: string;
  email: string;
  restaurantLegalName: string;
}

export type TermsFVWithBrowserInfo = TermsFormValues & { browserInfo: string };

interface SP {
  user: User;
  patchPending: boolean;
}

interface DP {
  clickOrInteract: (payload: ClickOrInteractPayload) => void;
}

interface OP {
  confirm: (formAndBrowserData: TermsFVWithBrowserInfo) => void;
}

type Props = SP & DP & OP;

export class TermsAndConditions extends Component<Props> {
  private addBrowserInfoAndConfirm: (formValues: TermsFormValues) => void;

  constructor(props: Props) {
    super(props);
    this.addBrowserInfoAndConfirm = (formValues: TermsFormValues) => {
      const platform = require('platform');
      const { name, version, layout, os, description }: any = platform;
      const browserInfo = JSON.stringify({ name, version, layout, os, description });
      const withBrowserInfo = { ...formValues, browserInfo, agreed: undefined };
      this.props.confirm(withBrowserInfo);
    };
  }

  public render() {
    const { user, patchPending, clickOrInteract } = this.props;
    const userEmail = user.userName;
    /* i18n nightmare next line */
    const userFullName = `${user.firstName} ${user.lastName}`;
    const initialValues = {
      agreed: false,
      email: userEmail,
      name: userFullName,
    };

    const clientHasRead = (
      <FormattedMessage
        id="gifts.terms.client_has_read"
        values={{
          'gifts-client-agreement-link': (
            <a href={LINK_TO_GIFTS_TERMS_V1} target="_blank">
              <FormattedMessage id="gifts.terms.gifts_client_agreement" />
            </a>
          ),
        }}
      />
    );

    return (
      <Form
        onSubmit={this.addBrowserInfoAndConfirm}
        initialValues={initialValues}
        validate={validateGiftsTermsForm}
        render={({ handleSubmit, active }) => (
          <form onSubmit={handleSubmit} className="terms-and-conditions-form gifts">
            <InputFieldAndLabel
              name="name"
              className="name"
              data-testid="name"
              messageIdSuffix="your_name"
              sublabelId="gifts.terms.must_be_authorized"
            />
            <div className="email">
              <FormattedMessage id="guestcampaigns.terms.your_email" tagName="label" />
              <span>{(initialValues as any).email}</span>
            </div>
            <InputFieldAndLabel
              name="restaurantLegalName"
              className="restaurant-legal-name"
              data-testid="restaurant-legal-name"
              messageIdSuffix="restaurant_legal_name"
            />
            <FinalFieldFocusTracker
              trackFocus={true}
              trackBlur={true}
              fieldName="restaurantLegalName"
              formattedFieldName="Restaurant Legal Name"
              active={active}
              clickOrInteract={clickOrInteract}
              pageName="Terms and Conditions modal"
            />
            <CheckboxFieldAndLabel name="agreed" className="agreed" label={clientHasRead as any} />
            <Button
              theme="publish"
              name="gifts-agree-and-sign-up"
              type="button"
              htmlType="submit"
              className="agree-and-publish"
              data-testid="agree-and-sign-up-button"
              busy={patchPending}
              onClick={() => {
                clickOrInteract({
                  action: 'Click',
                  props: { name: 'Publish', page: 'Terms and Conditions modal' },
                });
              }}
            >
              <FormattedMessage id="gifts.terms.sign_up" />
            </Button>
          </form>
        )}
      />
    );
  }
}

const mapState = (state: State) => ({
  patchPending: state.gifts.patchPending,
  user: selectors.user(state),
});

const mapDispatch = {
  clickOrInteract: actions.clickOrInteract,
};

export default connect<SP, DP, OP>(
  mapState,
  mapDispatch,
)(TermsAndConditions);

interface InputFieldAndLabelProps {
  name: string;
  className?: string;
  disabled?: boolean;
  messageIdSuffix: string;
  sublabelId?: string;
  onFocus?: () => void;
}
const InputFieldAndLabel: SFC<InputFieldAndLabelProps> = ({
  name,
  className,
  disabled,
  messageIdSuffix,
  sublabelId,
  onFocus,
}) => (
  <div className={className}>
    <FormattedMessage id={`guestcampaigns.terms.${messageIdSuffix}`} tagName="label" />
    <InputFinalField name={name} autoComplete="off" disabled={disabled || false} onFocus={onFocus} />
    {sublabelId && <FormattedMessage id={sublabelId} tagName="span" />}
  </div>
);

interface CheckboxAndLabelFieldProps {
  name: string;
  className?: string;
  label: string;
}
const CheckboxFieldAndLabel: SFC<CheckboxAndLabelFieldProps> = ({ name, className, label }: any) => {
  return <CheckboxFinalField name={name} className={className} label={label} />;
};
