// gc-microapp-reviews/src/store/reviewSummary/reviewSummaryDTO.ts
interface CategoryDistribution {
  Name: string;
  Count: number;
  Key: string;
}

interface DinersRecommendations {
  Count: number;
  Yes: number;
  No: number;
}

interface Overall {
  Rating: number;
  Count: number;
}

interface OverallRatingsDistribution {
  [rating: number]: number;
}

interface Ratings {
  OverallRatingsDistribution: OverallRatingsDistribution;
  DinersRecommendations: DinersRecommendations;
  Count: number;
  Overall: Overall;
  Food: Overall;
  Service: Overall;
  Ambience: Overall;
  Value: Overall;
  Noise: Overall;
}

interface RestaurantReviewSummary {
  Rid: number;
  Ratings: Ratings;
  RatingOnlyReviewRatings: Ratings;
  CategoryDistribution: CategoryDistribution[];
  TotalNumberOfReviews: number;
  AllTimeTextReviewCount: number;
  ShowReviews: boolean;
  StartDateUtc: string;
  Awards: any[];
  RestaurantName: string;
  Locale: string;
  LookBackDays: number;
  RatingBasedOn: string;
  VerifiedReviews: boolean;
}

export interface ReviewSummaryDTO {
  RestaurantReviewSummary: RestaurantReviewSummary;
}
