// gc-microapp-bistro/src/Restaurants/RidDiagnostic/RidDiff.tsx
import kebabCase from 'lodash/kebabCase';
import { propOr } from 'ramda';
import React from 'react';

import { Button } from '@buffet/button';

import { RidData } from '../../types';

import './RidDiff.scss';

interface RidDiffProps {
  primary: [ string, string, string ];
  secondary: [ string, string, string ];
  data: RidData[];
  onSync: () => void;
}

function getValue( data: RidData[], primary: [ string, string, string ]) {
  const source = data.find(d => d.title === primary[0]);
  if (!source) {
    return 'N/A';
  }
  // Convert all values to kebabCase to erase differences between styles from various sources.
  return kebabCase(propOr('notAvailable', primary[1], source.data) || 'notAvailable');
}

export function RidDiff({ data, primary, secondary, onSync }: RidDiffProps) {
  const primaryValue = getValue(data, primary);
  const secondaryValue = getValue(data, secondary);
  return data && primaryValue !== secondaryValue
  ? (
      <div className="data-diagnostic-result">
        <p className="warning">Product types don't match!</p>
        <div className="data-table">
          <div className="data-summary">
            {primary[0]} &middot; {primary[2]}: <strong>{primaryValue}</strong>
          </div>
          <div className="data-summary">
            {secondary[0]} &middot; {secondary[2]}: <strong>{secondaryValue}</strong>
          </div>
          <div className="data-summary">
            <Button theme="primary" type="button" onClick={onSync}>Sync {primary[0]} &#8646; {secondary[0]}</Button>
          </div>
        </div>
      </div>
    )
   : (null);
}
