// gc-microapp-custom-promotions/src/helpers/makeCampaignDTO.ts
import { User } from 'gca-react-components/src/redux-modules/currentContext';
import { CampaignState, CampaignTypes } from '../store/customPromotions/customPromotions.types';
import makeSchedule from './makeSchedule';
import { translateFromForm } from './suppression-dates-manipulator';

const makeCampaignDTO = (values: any, user: User, rid: any, approvalInfo: any, coverPrice: number) => {
  const { seatingDates: { endDate, startDate }, daysAndTimes, blockedDays, monthlyBudget: { cap, disabled } } = values;
  const { userName: email, firstName, lastName } = user;

  return {
    endDate: endDate === null ? undefined : endDate,
    schedule: makeSchedule(daysAndTimes),
    startDate,
    state: CampaignState.Live,
    suppressionDates: translateFromForm(blockedDays),
    type: CampaignTypes.CustomPromotions,

    campaignApprovalInfo: approvalInfo,
    changedByUser: email,
    maxBid: coverPrice,
    monthlyBudget: !disabled ? cap : undefined,
    restaurantContactEmail: email,
    /* i18n nightmare on next line */
    restaurantContactName: `${firstName} ${lastName}`,
    rid,
  };
};

export default makeCampaignDTO;
