// harness-smartling-client/src/services/SmartlingUploadService.js
module.exports = function (SmartlingBaseService, HTTPService, HTTPLogging, SmartlingConstants, Utils, Promise) {
  class SmartlingUploadService extends SmartlingBaseService {
    static send(options) {
      return new SmartlingUploadService().send(options);
    }

    send(options) {
      this.options = options;

      return Utils.promiseQueue([
        this.validateOptions.bind(this),
        this.makeRequest.bind(this)
      ]);
    }

    validateOptions() {
      return new Promise((resolve, reject) => {
        const options = this.options;

        if (!options.accessToken) {
          reject('Invalid Smartling Upload Option accessToken is missing');
        }

        if (!options.projectId) {
          reject('Invalid Smartling Upload Option projectId is missing');
        }

        if (!options.fileUri) {
          reject('Invalid Smartling Upload Option fileUri is missing');
        }

        if (!options.content) {
          reject('Invalid Smartling Upload Option content is missing');
        }

        if (!options.contentType) {
          reject('Invalid Smartling Upload Option contentType is missing');
        }

        resolve(options);
      });
    }

    makeRequest(options) {
      return HTTPService.post(this.getRequestUrl(options))
        .set('Authorization', `Bearer ${options.accessToken}`)
        .plugin(HTTPLogging)
        .appendFile('file', new Buffer(options.content), {
          filename: options.fileUri,
          contentType: options.contentType
        })
        .promiseBody()
        .then(this.validateResponse);
    }

    getRequestUrl(options) {
      const baseUrl = `/files-api/v2/projects/${options.projectId}/file`;
      const queryString = `?fileUri=${options.fileUri}&fileType=${options.fileType}`;
      return `${SmartlingConstants.host}${baseUrl}${queryString}`;
    }
  }

  return SmartlingUploadService;
};
