// user-agreements/src/store/api.ts
import { ajaxGetJSON, ajaxPatch, ajaxPost } from 'rxjs/observable/dom/AjaxObservable';
import './commonRxjs';
import { AcceptTermsDTO, AppointmentDateDTO, UpdateUserAgreementDTO } from './userAgreements/userAgreements.types';

const defaultAjaxOptions = { 'Content-Type': 'application/json' };

export function fetchUserAgreements(agreementId: string, validateUser: boolean) {
  if (validateUser) {
    return ajaxGetJSON(`/admin/api/bistro/user-agreements/${agreementId}`);
  } else {
    return ajaxGetJSON(`/admin/api/bistro/user-agreements/view/${agreementId}`);
  }
}

export function updateAgreementState(updateUserAgreement: UpdateUserAgreementDTO) {
  let data;

  if (updateUserAgreement.appointmentDate) {
    data = {
      agreementState: updateUserAgreement.agreementState,
      appointmentDate: updateUserAgreement.appointmentDate,
    };
  } else {
    data = {
      agreementState: updateUserAgreement.agreementState,
    };
  }

  return ajaxPatch(`/admin/api/bistro/user-agreements/${updateUserAgreement.agreementId}`,
    {
      data,
    },
    defaultAjaxOptions);
}

export function acceptTerms(acceptTermsPayload: AcceptTermsDTO) {
  const body = {
    clientEmail: acceptTermsPayload.clientEmail,
    signedVersions: acceptTermsPayload.signedVersions,
  };

  return ajaxPost(`/admin/api/bistro/user-agreements/${acceptTermsPayload.agreementId}`, body, defaultAjaxOptions);
}

export function appointmentScheduled(appointmentDatePayload: AppointmentDateDTO) {
  const body = {
    appointmentDate: appointmentDatePayload.appointmentDate,
  };

  return ajaxPost(
    `/admin/api/bistro/user-agreements/${appointmentDatePayload.agreementId}/appointment-scheduled`,
    body,
    defaultAjaxOptions,
  );
}
