// damper/src/components/SlideController/stories/barSlide.story.js
import React from 'react';
import { Graph } from '../../../visualisations/Graph';
import times from 'lodash.times';
import Chance from 'chance';

const padding = {
  bottom: 50,
  left: 50,
  right: 20,
  top: 20,
};

const largeSize = 7;

const customTheme = {
  axis: {
    x: {
      easing: {
        opacity: 'gentle',
      },
    },
    y: {
      stroke: {
        fill: 'transparent',
      },
    },
  },
  bars: {
    default: {
      rounding: 4,
      width: 40,
    },
  },
  common: {
    fillOpacity: 1,
    mouseOver: {
      fillOpacity: 0,
      strokeOpacity: 1,
    },
    strokeOpacity: 1,
  },
};

export const barStory = (columns) => {
  const chance = new Chance('Bar');

  const bars = columns.map((col) => {
    return {
      id: col.id,
      isInteractive: false,
      values: [
        {
          theme: {
            id: 'teal',
            style: {
              fill: '#76BCD8',
            },
          },
          value: chance.weighted(times(12, (n) => n), [2, 8, 4, 5, 3, 4, 2, 1, 2, 3, 2, 3]),
        },
        {
          theme: {
            id: 'yellow',
            style: {
              fill: '#F6C96B',
            },
          },
          value: chance.weighted(times(12, (n) => n), [2, 8, 4, 7, 6, 5, 2, 1, 2, 2, 1, 1]),
        }
      ],
    }
  });

  return (
    <Graph
      columns={columns}
      columnOffsets={{
        left: 0,
        right: 0,
      }}
      customTheme={customTheme}
      bars={bars}
      width={1010}
      height={500}
      padding={padding}
      responsive={'none'}
    />
  );
}
