// damper/src/components/SlideController/stories/icicleSlide.story.js
import React from 'react';
import { Icicle } from '../../../visualisations/Icicle';
import times from 'lodash.times';
import Chance from 'chance';

const padding = {
  bottom: 50,
  left: 50,
  right: 20,
  top: 20,
};

const largeSize = 7;

export const icicleStory = (columns) => {
  const chance = new Chance('Icicle');

  const data = columns.map((col) => {
    const partyCount = chance.weighted(times(12, (n) => n), [2, 8, 4, 7, 6, 5, 2, 1, 2, 2, 1, 1]);
    const parties = times(partyCount, () => {
      return chance.weighted(times(14, (n) => n), [
        3,
        10,
        4,
        7,
        6,
        5,
        6,
        2,
        2,
        3,
        1,
        1,
        1,
        1,
      ]);
    });

    return {
      id: col.id,
      values: parties,
    }
  });

  return (
    <Icicle
      columns={columns}
      columnOffsets={{
        left: 0,
        right: 0,
      }}
      data={data}
      width={1010}
      height={500}
      padding={padding}
      responsive={'none'}
      largeValue={largeSize}
    />
  );
}
