// gc-microapp-bistro/src/config.ts
// tslint:disable: object-literal-sort-keys
import { RawConfig } from './types';

const config: RawConfig[] = [
  {
    dataSource: 'ProductCheck',
    idType: 'rid',
    url: '/admin/api/bistro/restaurants/{rid}/products/diagnostic',
    dataItems: [
      { key: 'rid', label: 'RID' },
      { key: 'primaryProductMatching', label: 'Matching' },
      { key: 'rideProducts', label: 'Ride' },
      { key: 'otrcProducts', label: 'OTRC' },
      { key: 'webdbProduct', label: 'WebDB' },
      { key: 'ridePrimaryProduct', label: 'Ride Primary' },
      { key: 'otrcPrimaryProduct', label: 'OTRC Primary' },
      { key: 'webdbPrimaryProduct', label: 'WebDB Primary' },
    ],
  },
  {
    dataSource: 'SalesForce',
    idType: 'rid',
    url: '/admin/api/bistro/salesforce/{rid}',
    dataItems: [
      { key: 'message', label: 'Result', optional: true  },
      { key: 'statusCode', label: 'Status Code', optional: true },
      { key: 'rid', label: 'RID' },
      { key: 'accountId', label: 'Account ID' },
      { key: 'id', label: 'ID '},
      { key: 'productCode', label: 'Product Code' },
      { key: 'status', label: 'Status' },
      { key: 'lastModified', label: 'Modified on'},
    ],
  },
  {
    dataSource: 'RIDE',
    idType: 'rid',
    url: '/admin/api/bistro/ride/{rid}',
    dataItems: [
      { key: 'message', label: 'Result', optional: true  },
      { key: 'statusCode', label: 'Status Code', optional: true },
      { key: 'restaurantName', label: 'Name' },
      { key: 'accountStatus', label: 'Status'},
      { key: 'testRestaurant', label: 'Is Test Restaurant'},
      { key: 'salesforceId', label: 'Salesforce ID' },
      { key: 'romsId', label: 'ROMS ID' },
      { key: 'primaryProduct', label: 'Primary Product' },
      { key: 'supportedProducts', label: 'Supported Products' },
    ],
  },
  {
    dataSource: 'SideCar',
    idType: 'rid',
    url: '/admin/api/bistro/sidecar/{rid}',
    dataItems: [
      { key: 'message', label: 'Result', optional: true },
      { key: 'statusCode', label: 'Status Code', optional: true },
      { key: 'RID', label: 'RID' },
      { key: 'RestaurantType', label: 'Type' },
      { key: 'sidecarReporting.Restaurant.SidecarVersion', label: 'SideCar Version' },
      { key: 'OperationalStatus', label: 'Status' },
    ],
  },
  {
    dataSource: 'Restaurant Admin Service',
    idType: 'rid',
    url: '/admin/api/bistro/ras/{rid}',
    dataItems: [
      { key: 'message', label: 'Result', optional: true  },
      { key: 'statusCode', label: 'Status Code', optional: true },
      { key: 'core.restaurant.countryCode', label: 'Country Code'},
      { key: 'core.restaurant.restaurantType', label: 'Core Type' },
      { key: 'core.restaurant.restaurantId', label: 'Core ID' },
      { key: 'content.restaurantLocal.en-US.isActive', label: 'US - Active'},
      { key: 'content.restaurantLocal.en-US.restaurantName', label: 'US - Name'},
      { key: 'content.restaurantLocal.en-US.city', label: 'US - City'},
      { key: 'content.restaurantLocal.en-GB.isActive', label: 'GB - Active'},
      { key: 'content.restaurantLocal.en-GB.restaurantName', label: 'GB - Name'},
    ],
  },
  {
    dataSource: 'Venue Service',
    idType: 'rid',
    url: '/admin/api/bistro/venue/{rid}',
    dataItems: [
      { key: 'message', label: 'Result', optional: true  },
      { key: 'statusCode', label: 'Status Code', optional: true },
      { key: 'RID', label: 'RID' },
      { key: 'OperationalState', label: 'Status' },
    ],
  },
  {
    dataSource: 'GCAPS',
    idType: 'rid',
    url: 'setup/api/restaurant/{rid}',
    dataItems: [
      { key: 'message', label: 'Result' , optional: true },
      { key: 'statusCode', label: 'Status Code', optional: true },
      { key: 'pendingProductType', label: 'Pending Product Type' },
      { key: 'productType', label: 'Product Type' },
    ],
  },
  {
    dataSource: 'Orchestration',
    idType: 'rid',
    url: null,
    dataItems: [
      { key: 'message', label: 'Result' , optional: true },
      { key: 'statusCode', label: 'Status Code', optional: true },
    ],
  },
  {
    dataSource: 'OTRC',
    idType: 'roms',
    url: '/admin/api/bistro/otrc/{rom}',
    dataItems: [
      { key: 'message', label: 'Result', optional: true  },
      { key: 'statusCode', label: 'Status Code', optional: true },
      { key: 'restaurantName', label: 'Restaurant Name' },
      { key: 'rid', label: 'RID' },
      { key: 'romsId', label: 'ROMS ID' },
      { key: 'prodList', label: 'Product List' },
    ],
  },
  {
    dataSource: 'ImportStatus',
    idType: 'rid',
    url: '/admin/api/restaurant/{rid}/system-change/import-status',
    dataItems: [
      { key: 'message', label: 'Result', optional: true  },
      { key: 'statusCode', label: 'Status Code', optional: true },
      { key: 'data', label: 'Data', optional: true },
      { key: 'sidecarReporting.RID', label: 'RID' },
      { key: 'sidecarReporting.OperationalStatus', label: 'OperationalStatus' },
      { key: 'sidecarReporting.productType', label: 'Product Type' },
      { key: 'sidecarReporting.pendingProductType', label: 'Pending ProdType' },
      { key: 'sidecarReporting.RestaurantType', label: 'RestaurantType' },
      { key: 'venueSummary.ReservationCount', label: 'ReservationCount' },
      { key: 'venueSummary.GuestCount', label: 'GuestCount' },
    ],
  },
];

export default config;
