// gca-registry-publish/version-api/config/version-api-env-config.js
/*
*  Server information for the actual Microapp Version API
*  https://github.com/opentable/ot-microapp-version-api
*  This is a list of which environments are available to be configured
*  in the production and non-production instances of Microapp Version API
*
*  Used by tasks:  init-app, create-version
*/
const versionApiServers = {
  pp: {
    host: 'https://microapps-qa.otenv.com',
    environments: ['local', 'development', 'ci-sf', 'pp-sf', 'rc-sf'],
  },
  prod: {
    host: 'https://microapps.otenv.com',
    environments: ['stage-sc', 'prod-sc'],
  },
}

/*
*  Profiles are not environments, but rather a collection of environments,
*  bundled for developer convenience of executing a task across all of them.
*
*  Used by tasks:  promote-version
*/
const versionApiProfiles = {
  testing: [
    {
      name: 'local',
      host: versionApiServers.pp.host,
    },
    {
      name: 'development',
      host: versionApiServers.pp.host,
    },
    {
      name: 'ci-sf',
      host: versionApiServers.pp.host,
    },
    {
      name: 'pp-sf',
      host: versionApiServers.pp.host,
    },
    {
      name: 'rc-sf',
      host: versionApiServers.pp.host,
    },
    {
      name: 'stage-sc',
      host: versionApiServers.prod.host,
    },
  ],
  production: [
    {
      name: 'prod-sc',
      host: versionApiServers.prod.host,
    },
  ]
}

module.exports = {
  versionApiServers,
  versionApiProfiles,
}
