// gc-microapp-custom-promotions/src/CustomPromotions/BlockedDays/BlockedDatesPickerModal/BlockedDatesPickerModal.tsx
import React from 'react';
import { InjectedIntlProps, injectIntl } from 'react-intl';
import { convertSuppressionDatesToMap, translateMap } from '../../../helpers/suppression-dates-manipulator';
import { SuppressionDates } from '../../../store/customPromotions/customPromotions.types';
import Modal from '../../Modal/Modal';
import ModalHeader from '../../Modal/ModalHeader';
import BlockedDatesPicker, { DateAttributes } from './BlockedDatesPicker';

interface OP {
  suppressionDates: SuppressionDates;
  confirm: (payload: SuppressionDates) => void;
  close: (() => void);
}

type Props = OP & InjectedIntlProps;

interface ComponentState {
  suppressionDates: SuppressionDates;
}

/* P, S */
class BlockedDatesPickerModal extends React.Component<Props, ComponentState> {
  public constructor(props: any) {
    super(props);
    const { suppressionDates } = props;
    this.state = {
      suppressionDates,
    };
  }

  public render() {
    return (
      <Modal onClose={this.props.close} modalClass="edit-blocked-dates-modal">
        <ModalHeader
          onClose={this.props.close}
          title={this.props.intl.formatMessage({ id: 'guestcampaigns.custom_promotions.form.blocked_days.edit' })}
        />
        <BlockedDatesPicker
          blocked={convertSuppressionDatesToMap(this.state.suppressionDates)}
          setSuppressionDates={this.setSuppressionDates}
        />
      </Modal>
    );
  }

  public setSuppressionDates = (datesMap: Map<string, DateAttributes>) => {
    this.props.confirm(translateMap(datesMap));
    this.props.close();
  };
}

export default injectIntl(BlockedDatesPickerModal);
