// gca-jest-config/lib/jest.config.js
module.exports = {
  collectCoverageFrom: ['src/**/*.{ts,tsx}', '!src/translations/**/*.{js,json}', '!src/typings.d.ts'],
  moduleDirectories: ['node_modules', 'src'],
  moduleFileExtensions: ['ts', 'tsx', 'js', 'json'],
  moduleNameMapper: {
    'shared-react-utils(.*)$': '<rootDir>/../../shared_modules/shared-react-utils$1',
    'shared-utils(.*)$': '<rootDir>/../../shared_modules/shared-utils$1',
    'shared-react-atomic-ui(.*)$': '<rootDir>/../../shared_modules/shared-react-atomic-ui$1',
    'shared-react-smart-ui(.*)$': '<rootDir>/../../shared_modules/shared-react-smart-ui$1',
    '.*\\.(css|scss)$': 'gca-jest-config/lib/fake_css.js',
  },
  setupFiles: ['gca-jest-config/lib/test-bundler.js'],
  snapshotSerializers: ['enzyme-to-json/serializer'],
  testRegex: '\\.test\\.(ts|tsx|js|jsx)$',
  transform: {
    '\\.(ts|tsx)$': 'ts-jest/preprocessor.js',
    '^.+\\.(js|jsx)$': 'gca-babel-config',
    '\\.(jpg|svg|png|gif)$': 'gca-jest-config/lib/fake_path.js',
  },
  transformIgnorePatterns: ['node_modules/(?!gca-react-components|@buffet)'],
};
