// damper/src/visualisations/Graph/stories/area/helpers/Area.helpers.story.tsx
import { boolean, object, select } from '@storybook/addon-knobs';
import { tealLighter } from 'otkit-colors/token.common';
import React from 'react';
import { Graph } from '../../../Graph/index';
import ReadMe from './README.md';

const defaultColumns = [
  {
    id: 'a',
    label: 'A',
  },
  {
    id: 'b',
    label: 'B',
  },
  {
    id: 'c',
    label: 'C',
  },
  {
    id: 'd',
    label: 'D',
  },
  {
    id: 'e',
    label: 'E',
  },
];

const defaultAreas = [{
  data: [
    {
      data: [
        {
          id: 'a',
          value: 22,
        },
        {
          id: 'b',
          value: 122,
        },
        {
          id: 'c',
          value: 12.8,
        },
        {
          id: 'd',
          value: 10,
        },
        {
          id: 'e',
          value: 33,
        },
      ],
      id: 'area-0-1',
    },
  ],
  group: false,
  id: 'area-0',
},
{
  data: [
    {
      data: [
        {
          id: 'a',
          value: 122,
        },
        {
          id: 'b',
          value: 55,
        },
        {
          id: 'c',
          value: 328,
        },
        {
          id: 'd',
          value: 131,
        },
        {
          id: 'e',
          value: 133,
        },
      ],
      id: 'area-1-1',
      styles: {
        common: {
          fill: 'transparent',
          mouseOver: {
            fill: 'transparent',
            stroke: tealLighter,
          },
          stroke: tealLighter,
          strokeOpacity: 1,
          strokeWidth: 2,
        },
        helpers: {
          area: {
            points: {
              stroke: tealLighter,
            },
          },
        },
      },
    },
  ],
  group: false,
  id: 'area-1',
}];

const columnOffsets = {
  left: 0,
  right: 0,
};

const defaultPadding = {
  bottom: 40,
  left: 60,
  right: 60,
  top: 20,
};

export const areaHelpersStory = () => {
  const responsive = select('Responsive', ['redraw', 'scale', 'none'], 'redraw');
  const animation = select('Animation', ['all', 'none'], 'all');
  const isVertical = boolean('Show Vertical', false);
  const isHighlight = boolean('Show Highlight', false);
  const isPoints = boolean('Show Points', false);
  const columns = object('Columns', defaultColumns);
  const areasData = object('Area Data', defaultAreas);
  const helpers = {
    area: {
      highlight: {
        end: 'c',
        offset: {
          max: 'e',
          min: 'a',
        },
        show: isHighlight,
        start: 'c',
        width: -1,
      },
      points: {
        show: isPoints,
      },
      vertical: {
        show: isVertical,
      },
    },
  };

  return (
    <article className="markdown-body">
      <h1>Area</h1>
      <Graph
        width={1010}
        height={500}
        animation={animation}
        areas={areasData}
        columns={columns}
        padding={defaultPadding}
        responsive={responsive}
        columnOffsets={columnOffsets}
        helpers={helpers}
      />
    </article>
  );
};

export const areaHelpersStoryReadMe = ReadMe;
